// creem/src/models/operations/cancelsubscription.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */

import * as z from "zod";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { Result as SafeParseResult } from "../../types/fp.js";
import { SDKValidationError } from "../errors/sdkvalidationerror.js";

export type CancelSubscriptionRequest = {
  id: string;
  xApiKey: string;
};

/** @internal */
export const CancelSubscriptionRequest$inboundSchema: z.ZodType<
  CancelSubscriptionRequest,
  z.ZodTypeDef,
  unknown
> = z.object({
  id: z.string(),
  "x-api-key": z.string(),
}).transform((v) => {
  return remap$(v, {
    "x-api-key": "xApiKey",
  });
});

/** @internal */
export type CancelSubscriptionRequest$Outbound = {
  id: string;
  "x-api-key": string;
};

/** @internal */
export const CancelSubscriptionRequest$outboundSchema: z.ZodType<
  CancelSubscriptionRequest$Outbound,
  z.ZodTypeDef,
  CancelSubscriptionRequest
> = z.object({
  id: z.string(),
  xApiKey: z.string(),
}).transform((v) => {
  return remap$(v, {
    xApiKey: "x-api-key",
  });
});

/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
export namespace CancelSubscriptionRequest$ {
  /** @deprecated use `CancelSubscriptionRequest$inboundSchema` instead. */
  export const inboundSchema = CancelSubscriptionRequest$inboundSchema;
  /** @deprecated use `CancelSubscriptionRequest$outboundSchema` instead. */
  export const outboundSchema = CancelSubscriptionRequest$outboundSchema;
  /** @deprecated use `CancelSubscriptionRequest$Outbound` instead. */
  export type Outbound = CancelSubscriptionRequest$Outbound;
}

export function cancelSubscriptionRequestToJSON(
  cancelSubscriptionRequest: CancelSubscriptionRequest,
): string {
  return JSON.stringify(
    CancelSubscriptionRequest$outboundSchema.parse(cancelSubscriptionRequest),
  );
}

export function cancelSubscriptionRequestFromJSON(
  jsonString: string,
): SafeParseResult<CancelSubscriptionRequest, SDKValidationError> {
  return safeParse(
    jsonString,
    (x) => CancelSubscriptionRequest$inboundSchema.parse(JSON.parse(x)),
    `Failed to parse 'CancelSubscriptionRequest' from JSON`,
  );
}
