// creem/src/models/operations/deactivatelicense.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */

import * as z from "zod";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { Result as SafeParseResult } from "../../types/fp.js";
import * as components from "../components/index.js";
import { SDKValidationError } from "../errors/sdkvalidationerror.js";

export type DeactivateLicenseRequest = {
  xApiKey: string;
  deactivateLicenseRequestEntity: components.DeactivateLicenseRequestEntity;
};

/** @internal */
export const DeactivateLicenseRequest$inboundSchema: z.ZodType<
  DeactivateLicenseRequest,
  z.ZodTypeDef,
  unknown
> = z.object({
  "x-api-key": z.string(),
  DeactivateLicenseRequestEntity:
    components.DeactivateLicenseRequestEntity$inboundSchema,
}).transform((v) => {
  return remap$(v, {
    "x-api-key": "xApiKey",
    "DeactivateLicenseRequestEntity": "deactivateLicenseRequestEntity",
  });
});

/** @internal */
export type DeactivateLicenseRequest$Outbound = {
  "x-api-key": string;
  DeactivateLicenseRequestEntity:
    components.DeactivateLicenseRequestEntity$Outbound;
};

/** @internal */
export const DeactivateLicenseRequest$outboundSchema: z.ZodType<
  DeactivateLicenseRequest$Outbound,
  z.ZodTypeDef,
  DeactivateLicenseRequest
> = z.object({
  xApiKey: z.string(),
  deactivateLicenseRequestEntity:
    components.DeactivateLicenseRequestEntity$outboundSchema,
}).transform((v) => {
  return remap$(v, {
    xApiKey: "x-api-key",
    deactivateLicenseRequestEntity: "DeactivateLicenseRequestEntity",
  });
});

/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
export namespace DeactivateLicenseRequest$ {
  /** @deprecated use `DeactivateLicenseRequest$inboundSchema` instead. */
  export const inboundSchema = DeactivateLicenseRequest$inboundSchema;
  /** @deprecated use `DeactivateLicenseRequest$outboundSchema` instead. */
  export const outboundSchema = DeactivateLicenseRequest$outboundSchema;
  /** @deprecated use `DeactivateLicenseRequest$Outbound` instead. */
  export type Outbound = DeactivateLicenseRequest$Outbound;
}

export function deactivateLicenseRequestToJSON(
  deactivateLicenseRequest: DeactivateLicenseRequest,
): string {
  return JSON.stringify(
    DeactivateLicenseRequest$outboundSchema.parse(deactivateLicenseRequest),
  );
}

export function deactivateLicenseRequestFromJSON(
  jsonString: string,
): SafeParseResult<DeactivateLicenseRequest, SDKValidationError> {
  return safeParse(
    jsonString,
    (x) => DeactivateLicenseRequest$inboundSchema.parse(JSON.parse(x)),
    `Failed to parse 'DeactivateLicenseRequest' from JSON`,
  );
}
