// creem/src/models/operations/retrievecheckout.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */

import * as z from "zod";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { Result as SafeParseResult } from "../../types/fp.js";
import { SDKValidationError } from "../errors/sdkvalidationerror.js";

export type RetrieveCheckoutRequest = {
  checkoutId: string;
  xApiKey: string;
};

/** @internal */
export const RetrieveCheckoutRequest$inboundSchema: z.ZodType<
  RetrieveCheckoutRequest,
  z.ZodTypeDef,
  unknown
> = z.object({
  checkout_id: z.string(),
  "x-api-key": z.string(),
}).transform((v) => {
  return remap$(v, {
    "checkout_id": "checkoutId",
    "x-api-key": "xApiKey",
  });
});

/** @internal */
export type RetrieveCheckoutRequest$Outbound = {
  checkout_id: string;
  "x-api-key": string;
};

/** @internal */
export const RetrieveCheckoutRequest$outboundSchema: z.ZodType<
  RetrieveCheckoutRequest$Outbound,
  z.ZodTypeDef,
  RetrieveCheckoutRequest
> = z.object({
  checkoutId: z.string(),
  xApiKey: z.string(),
}).transform((v) => {
  return remap$(v, {
    checkoutId: "checkout_id",
    xApiKey: "x-api-key",
  });
});

/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
export namespace RetrieveCheckoutRequest$ {
  /** @deprecated use `RetrieveCheckoutRequest$inboundSchema` instead. */
  export const inboundSchema = RetrieveCheckoutRequest$inboundSchema;
  /** @deprecated use `RetrieveCheckoutRequest$outboundSchema` instead. */
  export const outboundSchema = RetrieveCheckoutRequest$outboundSchema;
  /** @deprecated use `RetrieveCheckoutRequest$Outbound` instead. */
  export type Outbound = RetrieveCheckoutRequest$Outbound;
}

export function retrieveCheckoutRequestToJSON(
  retrieveCheckoutRequest: RetrieveCheckoutRequest,
): string {
  return JSON.stringify(
    RetrieveCheckoutRequest$outboundSchema.parse(retrieveCheckoutRequest),
  );
}

export function retrieveCheckoutRequestFromJSON(
  jsonString: string,
): SafeParseResult<RetrieveCheckoutRequest, SDKValidationError> {
  return safeParse(
    jsonString,
    (x) => RetrieveCheckoutRequest$inboundSchema.parse(JSON.parse(x)),
    `Failed to parse 'RetrieveCheckoutRequest' from JSON`,
  );
}
