// creem/src/models/operations/searchproducts.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */

import * as z from "zod";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { Result as SafeParseResult } from "../../types/fp.js";
import { SDKValidationError } from "../errors/sdkvalidationerror.js";

export type SearchProductsRequest = {
  /**
   * The page number
   */
  pageNumber?: number | undefined;
  /**
   * The the page size
   */
  pageSize?: number | undefined;
  xApiKey: string;
};

/** @internal */
export const SearchProductsRequest$inboundSchema: z.ZodType<
  SearchProductsRequest,
  z.ZodTypeDef,
  unknown
> = z.object({
  page_number: z.number().optional(),
  page_size: z.number().optional(),
  "x-api-key": z.string(),
}).transform((v) => {
  return remap$(v, {
    "page_number": "pageNumber",
    "page_size": "pageSize",
    "x-api-key": "xApiKey",
  });
});

/** @internal */
export type SearchProductsRequest$Outbound = {
  page_number?: number | undefined;
  page_size?: number | undefined;
  "x-api-key": string;
};

/** @internal */
export const SearchProductsRequest$outboundSchema: z.ZodType<
  SearchProductsRequest$Outbound,
  z.ZodTypeDef,
  SearchProductsRequest
> = z.object({
  pageNumber: z.number().optional(),
  pageSize: z.number().optional(),
  xApiKey: z.string(),
}).transform((v) => {
  return remap$(v, {
    pageNumber: "page_number",
    pageSize: "page_size",
    xApiKey: "x-api-key",
  });
});

/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
export namespace SearchProductsRequest$ {
  /** @deprecated use `SearchProductsRequest$inboundSchema` instead. */
  export const inboundSchema = SearchProductsRequest$inboundSchema;
  /** @deprecated use `SearchProductsRequest$outboundSchema` instead. */
  export const outboundSchema = SearchProductsRequest$outboundSchema;
  /** @deprecated use `SearchProductsRequest$Outbound` instead. */
  export type Outbound = SearchProductsRequest$Outbound;
}

export function searchProductsRequestToJSON(
  searchProductsRequest: SearchProductsRequest,
): string {
  return JSON.stringify(
    SearchProductsRequest$outboundSchema.parse(searchProductsRequest),
  );
}

export function searchProductsRequestFromJSON(
  jsonString: string,
): SafeParseResult<SearchProductsRequest, SDKValidationError> {
  return safeParse(
    jsonString,
    (x) => SearchProductsRequest$inboundSchema.parse(JSON.parse(x)),
    `Failed to parse 'SearchProductsRequest' from JSON`,
  );
}
