// creem/src/models/operations/validatelicense.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */

import * as z from "zod";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { Result as SafeParseResult } from "../../types/fp.js";
import * as components from "../components/index.js";
import { SDKValidationError } from "../errors/sdkvalidationerror.js";

export type ValidateLicenseRequest = {
  xApiKey: string;
  validateLicenseRequestEntity: components.ValidateLicenseRequestEntity;
};

/** @internal */
export const ValidateLicenseRequest$inboundSchema: z.ZodType<
  ValidateLicenseRequest,
  z.ZodTypeDef,
  unknown
> = z.object({
  "x-api-key": z.string(),
  ValidateLicenseRequestEntity:
    components.ValidateLicenseRequestEntity$inboundSchema,
}).transform((v) => {
  return remap$(v, {
    "x-api-key": "xApiKey",
    "ValidateLicenseRequestEntity": "validateLicenseRequestEntity",
  });
});

/** @internal */
export type ValidateLicenseRequest$Outbound = {
  "x-api-key": string;
  ValidateLicenseRequestEntity:
    components.ValidateLicenseRequestEntity$Outbound;
};

/** @internal */
export const ValidateLicenseRequest$outboundSchema: z.ZodType<
  ValidateLicenseRequest$Outbound,
  z.ZodTypeDef,
  ValidateLicenseRequest
> = z.object({
  xApiKey: z.string(),
  validateLicenseRequestEntity:
    components.ValidateLicenseRequestEntity$outboundSchema,
}).transform((v) => {
  return remap$(v, {
    xApiKey: "x-api-key",
    validateLicenseRequestEntity: "ValidateLicenseRequestEntity",
  });
});

/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
export namespace ValidateLicenseRequest$ {
  /** @deprecated use `ValidateLicenseRequest$inboundSchema` instead. */
  export const inboundSchema = ValidateLicenseRequest$inboundSchema;
  /** @deprecated use `ValidateLicenseRequest$outboundSchema` instead. */
  export const outboundSchema = ValidateLicenseRequest$outboundSchema;
  /** @deprecated use `ValidateLicenseRequest$Outbound` instead. */
  export type Outbound = ValidateLicenseRequest$Outbound;
}

export function validateLicenseRequestToJSON(
  validateLicenseRequest: ValidateLicenseRequest,
): string {
  return JSON.stringify(
    ValidateLicenseRequest$outboundSchema.parse(validateLicenseRequest),
  );
}

export function validateLicenseRequestFromJSON(
  jsonString: string,
): SafeParseResult<ValidateLicenseRequest, SDKValidationError> {
  return safeParse(
    jsonString,
    (x) => ValidateLicenseRequest$inboundSchema.parse(JSON.parse(x)),
    `Failed to parse 'ValidateLicenseRequest' from JSON`,
  );
}
