// creem/dist/commonjs/models/components/creatediscountrequestentity.js
"use strict";
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateDiscountRequestEntity$ = exports.CreateDiscountRequestEntity$outboundSchema = exports.CreateDiscountRequestEntity$inboundSchema = exports.CreateDiscountRequestEntityDuration$ = exports.CreateDiscountRequestEntityDuration$outboundSchema = exports.CreateDiscountRequestEntityDuration$inboundSchema = exports.CreateDiscountRequestEntityType$ = exports.CreateDiscountRequestEntityType$outboundSchema = exports.CreateDiscountRequestEntityType$inboundSchema = exports.CreateDiscountRequestEntityDuration = exports.CreateDiscountRequestEntityType = void 0;
exports.createDiscountRequestEntityToJSON = createDiscountRequestEntityToJSON;
exports.createDiscountRequestEntityFromJSON = createDiscountRequestEntityFromJSON;
const z = __importStar(require("zod"));
const primitives_js_1 = require("../../lib/primitives.js");
const schemas_js_1 = require("../../lib/schemas.js");
/**
 * The type of the discount, either "percentage" or "fixed".
 */
exports.CreateDiscountRequestEntityType = {
    Percentage: "percentage",
    Fixed: "fixed",
};
/**
 * The duration type for the discount.
 */
exports.CreateDiscountRequestEntityDuration = {
    Forever: "forever",
    Once: "once",
    Repeating: "repeating",
};
/** @internal */
exports.CreateDiscountRequestEntityType$inboundSchema = z.nativeEnum(exports.CreateDiscountRequestEntityType);
/** @internal */
exports.CreateDiscountRequestEntityType$outboundSchema = exports.CreateDiscountRequestEntityType$inboundSchema;
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
var CreateDiscountRequestEntityType$;
(function (CreateDiscountRequestEntityType$) {
    /** @deprecated use `CreateDiscountRequestEntityType$inboundSchema` instead. */
    CreateDiscountRequestEntityType$.inboundSchema = exports.CreateDiscountRequestEntityType$inboundSchema;
    /** @deprecated use `CreateDiscountRequestEntityType$outboundSchema` instead. */
    CreateDiscountRequestEntityType$.outboundSchema = exports.CreateDiscountRequestEntityType$outboundSchema;
})(CreateDiscountRequestEntityType$ || (exports.CreateDiscountRequestEntityType$ = CreateDiscountRequestEntityType$ = {}));
/** @internal */
exports.CreateDiscountRequestEntityDuration$inboundSchema = z.nativeEnum(exports.CreateDiscountRequestEntityDuration);
/** @internal */
exports.CreateDiscountRequestEntityDuration$outboundSchema = exports.CreateDiscountRequestEntityDuration$inboundSchema;
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
var CreateDiscountRequestEntityDuration$;
(function (CreateDiscountRequestEntityDuration$) {
    /** @deprecated use `CreateDiscountRequestEntityDuration$inboundSchema` instead. */
    CreateDiscountRequestEntityDuration$.inboundSchema = exports.CreateDiscountRequestEntityDuration$inboundSchema;
    /** @deprecated use `CreateDiscountRequestEntityDuration$outboundSchema` instead. */
    CreateDiscountRequestEntityDuration$.outboundSchema = exports.CreateDiscountRequestEntityDuration$outboundSchema;
})(CreateDiscountRequestEntityDuration$ || (exports.CreateDiscountRequestEntityDuration$ = CreateDiscountRequestEntityDuration$ = {}));
/** @internal */
exports.CreateDiscountRequestEntity$inboundSchema = z.object({
    name: z.string(),
    code: z.string().optional(),
    type: exports.CreateDiscountRequestEntityType$inboundSchema,
    amount: z.number().optional(),
    currency: z.string().optional(),
    percentage: z.number().optional(),
    expiry_date: z.string().datetime({ offset: true }).transform(v => new Date(v))
        .optional(),
    max_redemptions: z.number().optional(),
    duration: exports.CreateDiscountRequestEntityDuration$inboundSchema,
    duration_in_months: z.number().optional(),
    applies_to_products: z.array(z.string()),
}).transform((v) => {
    return (0, primitives_js_1.remap)(v, {
        "expiry_date": "expiryDate",
        "max_redemptions": "maxRedemptions",
        "duration_in_months": "durationInMonths",
        "applies_to_products": "appliesToProducts",
    });
});
/** @internal */
exports.CreateDiscountRequestEntity$outboundSchema = z.object({
    name: z.string(),
    code: z.string().optional(),
    type: exports.CreateDiscountRequestEntityType$outboundSchema,
    amount: z.number().optional(),
    currency: z.string().optional(),
    percentage: z.number().optional(),
    expiryDate: z.date().transform(v => v.toISOString()).optional(),
    maxRedemptions: z.number().optional(),
    duration: exports.CreateDiscountRequestEntityDuration$outboundSchema,
    durationInMonths: z.number().optional(),
    appliesToProducts: z.array(z.string()),
}).transform((v) => {
    return (0, primitives_js_1.remap)(v, {
        expiryDate: "expiry_date",
        maxRedemptions: "max_redemptions",
        durationInMonths: "duration_in_months",
        appliesToProducts: "applies_to_products",
    });
});
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
var CreateDiscountRequestEntity$;
(function (CreateDiscountRequestEntity$) {
    /** @deprecated use `CreateDiscountRequestEntity$inboundSchema` instead. */
    CreateDiscountRequestEntity$.inboundSchema = exports.CreateDiscountRequestEntity$inboundSchema;
    /** @deprecated use `CreateDiscountRequestEntity$outboundSchema` instead. */
    CreateDiscountRequestEntity$.outboundSchema = exports.CreateDiscountRequestEntity$outboundSchema;
})(CreateDiscountRequestEntity$ || (exports.CreateDiscountRequestEntity$ = CreateDiscountRequestEntity$ = {}));
function createDiscountRequestEntityToJSON(createDiscountRequestEntity) {
    return JSON.stringify(exports.CreateDiscountRequestEntity$outboundSchema.parse(createDiscountRequestEntity));
}
function createDiscountRequestEntityFromJSON(jsonString) {
    return (0, schemas_js_1.safeParse)(jsonString, (x) => exports.CreateDiscountRequestEntity$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'CreateDiscountRequestEntity' from JSON`);
}
