// creem/dist/commonjs/models/components/customerentity.js
"use strict";
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomerEntity$ = exports.CustomerEntity$outboundSchema = exports.CustomerEntity$inboundSchema = exports.CustomerEntityMode$ = exports.CustomerEntityMode$outboundSchema = exports.CustomerEntityMode$inboundSchema = exports.CustomerEntityMode = void 0;
exports.customerEntityToJSON = customerEntityToJSON;
exports.customerEntityFromJSON = customerEntityFromJSON;
const z = __importStar(require("zod"));
const primitives_js_1 = require("../../lib/primitives.js");
const schemas_js_1 = require("../../lib/schemas.js");
/**
 * String representing the environment.
 */
exports.CustomerEntityMode = {
    Test: "test",
    Prod: "prod",
    Sandbox: "sandbox",
};
/** @internal */
exports.CustomerEntityMode$inboundSchema = z.nativeEnum(exports.CustomerEntityMode);
/** @internal */
exports.CustomerEntityMode$outboundSchema = exports.CustomerEntityMode$inboundSchema;
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
var CustomerEntityMode$;
(function (CustomerEntityMode$) {
    /** @deprecated use `CustomerEntityMode$inboundSchema` instead. */
    CustomerEntityMode$.inboundSchema = exports.CustomerEntityMode$inboundSchema;
    /** @deprecated use `CustomerEntityMode$outboundSchema` instead. */
    CustomerEntityMode$.outboundSchema = exports.CustomerEntityMode$outboundSchema;
})(CustomerEntityMode$ || (exports.CustomerEntityMode$ = CustomerEntityMode$ = {}));
/** @internal */
exports.CustomerEntity$inboundSchema = z.object({
    id: z.string(),
    mode: exports.CustomerEntityMode$inboundSchema,
    object: z.string(),
    email: z.string(),
    name: z.string().optional(),
    country: z.string(),
    created_at: z.string().datetime({ offset: true }).transform(v => new Date(v)),
    updated_at: z.string().datetime({ offset: true }).transform(v => new Date(v)),
}).transform((v) => {
    return (0, primitives_js_1.remap)(v, {
        "created_at": "createdAt",
        "updated_at": "updatedAt",
    });
});
/** @internal */
exports.CustomerEntity$outboundSchema = z.object({
    id: z.string(),
    mode: exports.CustomerEntityMode$outboundSchema,
    object: z.string(),
    email: z.string(),
    name: z.string().optional(),
    country: z.string(),
    createdAt: z.date().transform(v => v.toISOString()),
    updatedAt: z.date().transform(v => v.toISOString()),
}).transform((v) => {
    return (0, primitives_js_1.remap)(v, {
        createdAt: "created_at",
        updatedAt: "updated_at",
    });
});
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
var CustomerEntity$;
(function (CustomerEntity$) {
    /** @deprecated use `CustomerEntity$inboundSchema` instead. */
    CustomerEntity$.inboundSchema = exports.CustomerEntity$inboundSchema;
    /** @deprecated use `CustomerEntity$outboundSchema` instead. */
    CustomerEntity$.outboundSchema = exports.CustomerEntity$outboundSchema;
})(CustomerEntity$ || (exports.CustomerEntity$ = CustomerEntity$ = {}));
function customerEntityToJSON(customerEntity) {
    return JSON.stringify(exports.CustomerEntity$outboundSchema.parse(customerEntity));
}
function customerEntityFromJSON(jsonString) {
    return (0, schemas_js_1.safeParse)(jsonString, (x) => exports.CustomerEntity$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'CustomerEntity' from JSON`);
}
