// creem/dist/commonjs/models/components/deactivatelicenserequestentity.js
"use strict";
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeactivateLicenseRequestEntity$ = exports.DeactivateLicenseRequestEntity$outboundSchema = exports.DeactivateLicenseRequestEntity$inboundSchema = void 0;
exports.deactivateLicenseRequestEntityToJSON = deactivateLicenseRequestEntityToJSON;
exports.deactivateLicenseRequestEntityFromJSON = deactivateLicenseRequestEntityFromJSON;
const z = __importStar(require("zod"));
const primitives_js_1 = require("../../lib/primitives.js");
const schemas_js_1 = require("../../lib/schemas.js");
/** @internal */
exports.DeactivateLicenseRequestEntity$inboundSchema = z.object({
    key: z.string(),
    instance_id: z.string(),
}).transform((v) => {
    return (0, primitives_js_1.remap)(v, {
        "instance_id": "instanceId",
    });
});
/** @internal */
exports.DeactivateLicenseRequestEntity$outboundSchema = z.object({
    key: z.string(),
    instanceId: z.string(),
}).transform((v) => {
    return (0, primitives_js_1.remap)(v, {
        instanceId: "instance_id",
    });
});
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
var DeactivateLicenseRequestEntity$;
(function (DeactivateLicenseRequestEntity$) {
    /** @deprecated use `DeactivateLicenseRequestEntity$inboundSchema` instead. */
    DeactivateLicenseRequestEntity$.inboundSchema = exports.DeactivateLicenseRequestEntity$inboundSchema;
    /** @deprecated use `DeactivateLicenseRequestEntity$outboundSchema` instead. */
    DeactivateLicenseRequestEntity$.outboundSchema = exports.DeactivateLicenseRequestEntity$outboundSchema;
})(DeactivateLicenseRequestEntity$ || (exports.DeactivateLicenseRequestEntity$ = DeactivateLicenseRequestEntity$ = {}));
function deactivateLicenseRequestEntityToJSON(deactivateLicenseRequestEntity) {
    return JSON.stringify(exports.DeactivateLicenseRequestEntity$outboundSchema.parse(deactivateLicenseRequestEntity));
}
function deactivateLicenseRequestEntityFromJSON(jsonString) {
    return (0, schemas_js_1.safeParse)(jsonString, (x) => exports.DeactivateLicenseRequestEntity$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'DeactivateLicenseRequestEntity' from JSON`);
}
