// creem/dist/commonjs/models/components/discountentity.js
"use strict";
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.DiscountEntity$ = exports.DiscountEntity$outboundSchema = exports.DiscountEntity$inboundSchema = exports.Duration$ = exports.Duration$outboundSchema = exports.Duration$inboundSchema = exports.DiscountEntityType$ = exports.DiscountEntityType$outboundSchema = exports.DiscountEntityType$inboundSchema = exports.DiscountEntityStatus$ = exports.DiscountEntityStatus$outboundSchema = exports.DiscountEntityStatus$inboundSchema = exports.DiscountEntityMode$ = exports.DiscountEntityMode$outboundSchema = exports.DiscountEntityMode$inboundSchema = exports.Duration = exports.DiscountEntityType = exports.DiscountEntityStatus = exports.DiscountEntityMode = void 0;
exports.discountEntityToJSON = discountEntityToJSON;
exports.discountEntityFromJSON = discountEntityFromJSON;
const z = __importStar(require("zod"));
const primitives_js_1 = require("../../lib/primitives.js");
const schemas_js_1 = require("../../lib/schemas.js");
/**
 * String representing the environment.
 */
exports.DiscountEntityMode = {
    Test: "test",
    Prod: "prod",
    Sandbox: "sandbox",
};
/**
 * The status of the discount (e.g., active, inactive).
 */
exports.DiscountEntityStatus = {
    Active: "active",
    Draft: "draft",
    Expired: "expired",
    Scheduled: "scheduled",
};
/**
 * The type of the discount, either "percentage" or "fixed".
 */
exports.DiscountEntityType = {
    Percentage: "percentage",
    Fixed: "fixed",
};
/**
 * The duration type for the discount.
 */
exports.Duration = {
    Forever: "forever",
    Once: "once",
    Repeating: "repeating",
};
/** @internal */
exports.DiscountEntityMode$inboundSchema = z.nativeEnum(exports.DiscountEntityMode);
/** @internal */
exports.DiscountEntityMode$outboundSchema = exports.DiscountEntityMode$inboundSchema;
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
var DiscountEntityMode$;
(function (DiscountEntityMode$) {
    /** @deprecated use `DiscountEntityMode$inboundSchema` instead. */
    DiscountEntityMode$.inboundSchema = exports.DiscountEntityMode$inboundSchema;
    /** @deprecated use `DiscountEntityMode$outboundSchema` instead. */
    DiscountEntityMode$.outboundSchema = exports.DiscountEntityMode$outboundSchema;
})(DiscountEntityMode$ || (exports.DiscountEntityMode$ = DiscountEntityMode$ = {}));
/** @internal */
exports.DiscountEntityStatus$inboundSchema = z.nativeEnum(exports.DiscountEntityStatus);
/** @internal */
exports.DiscountEntityStatus$outboundSchema = exports.DiscountEntityStatus$inboundSchema;
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
var DiscountEntityStatus$;
(function (DiscountEntityStatus$) {
    /** @deprecated use `DiscountEntityStatus$inboundSchema` instead. */
    DiscountEntityStatus$.inboundSchema = exports.DiscountEntityStatus$inboundSchema;
    /** @deprecated use `DiscountEntityStatus$outboundSchema` instead. */
    DiscountEntityStatus$.outboundSchema = exports.DiscountEntityStatus$outboundSchema;
})(DiscountEntityStatus$ || (exports.DiscountEntityStatus$ = DiscountEntityStatus$ = {}));
/** @internal */
exports.DiscountEntityType$inboundSchema = z.nativeEnum(exports.DiscountEntityType);
/** @internal */
exports.DiscountEntityType$outboundSchema = exports.DiscountEntityType$inboundSchema;
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
var DiscountEntityType$;
(function (DiscountEntityType$) {
    /** @deprecated use `DiscountEntityType$inboundSchema` instead. */
    DiscountEntityType$.inboundSchema = exports.DiscountEntityType$inboundSchema;
    /** @deprecated use `DiscountEntityType$outboundSchema` instead. */
    DiscountEntityType$.outboundSchema = exports.DiscountEntityType$outboundSchema;
})(DiscountEntityType$ || (exports.DiscountEntityType$ = DiscountEntityType$ = {}));
/** @internal */
exports.Duration$inboundSchema = z
    .nativeEnum(exports.Duration);
/** @internal */
exports.Duration$outboundSchema = exports.Duration$inboundSchema;
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
var Duration$;
(function (Duration$) {
    /** @deprecated use `Duration$inboundSchema` instead. */
    Duration$.inboundSchema = exports.Duration$inboundSchema;
    /** @deprecated use `Duration$outboundSchema` instead. */
    Duration$.outboundSchema = exports.Duration$outboundSchema;
})(Duration$ || (exports.Duration$ = Duration$ = {}));
/** @internal */
exports.DiscountEntity$inboundSchema = z.object({
    id: z.string(),
    mode: exports.DiscountEntityMode$inboundSchema,
    object: z.string(),
    status: exports.DiscountEntityStatus$inboundSchema,
    name: z.string(),
    code: z.string(),
    type: exports.DiscountEntityType$inboundSchema,
    amount: z.number().optional(),
    currency: z.string().optional(),
    percentage: z.number().optional(),
    expiry_date: z.string().datetime({ offset: true }).transform(v => new Date(v))
        .optional(),
    max_redemptions: z.number().optional(),
    duration: exports.Duration$inboundSchema.optional(),
    duration_in_months: z.number().optional(),
    applies_to_products: z.array(z.string()).optional(),
}).transform((v) => {
    return (0, primitives_js_1.remap)(v, {
        "expiry_date": "expiryDate",
        "max_redemptions": "maxRedemptions",
        "duration_in_months": "durationInMonths",
        "applies_to_products": "appliesToProducts",
    });
});
/** @internal */
exports.DiscountEntity$outboundSchema = z.object({
    id: z.string(),
    mode: exports.DiscountEntityMode$outboundSchema,
    object: z.string(),
    status: exports.DiscountEntityStatus$outboundSchema,
    name: z.string(),
    code: z.string(),
    type: exports.DiscountEntityType$outboundSchema,
    amount: z.number().optional(),
    currency: z.string().optional(),
    percentage: z.number().optional(),
    expiryDate: z.date().transform(v => v.toISOString()).optional(),
    maxRedemptions: z.number().optional(),
    duration: exports.Duration$outboundSchema.optional(),
    durationInMonths: z.number().optional(),
    appliesToProducts: z.array(z.string()).optional(),
}).transform((v) => {
    return (0, primitives_js_1.remap)(v, {
        expiryDate: "expiry_date",
        maxRedemptions: "max_redemptions",
        durationInMonths: "duration_in_months",
        appliesToProducts: "applies_to_products",
    });
});
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
var DiscountEntity$;
(function (DiscountEntity$) {
    /** @deprecated use `DiscountEntity$inboundSchema` instead. */
    DiscountEntity$.inboundSchema = exports.DiscountEntity$inboundSchema;
    /** @deprecated use `DiscountEntity$outboundSchema` instead. */
    DiscountEntity$.outboundSchema = exports.DiscountEntity$outboundSchema;
})(DiscountEntity$ || (exports.DiscountEntity$ = DiscountEntity$ = {}));
function discountEntityToJSON(discountEntity) {
    return JSON.stringify(exports.DiscountEntity$outboundSchema.parse(discountEntity));
}
function discountEntityFromJSON(jsonString) {
    return (0, schemas_js_1.safeParse)(jsonString, (x) => exports.DiscountEntity$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'DiscountEntity' from JSON`);
}
