// creem/dist/commonjs/models/components/paginationentity.js
"use strict";
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.PaginationEntity$ = exports.PaginationEntity$outboundSchema = exports.PaginationEntity$inboundSchema = void 0;
exports.paginationEntityToJSON = paginationEntityToJSON;
exports.paginationEntityFromJSON = paginationEntityFromJSON;
const z = __importStar(require("zod"));
const primitives_js_1 = require("../../lib/primitives.js");
const schemas_js_1 = require("../../lib/schemas.js");
/** @internal */
exports.PaginationEntity$inboundSchema = z.object({
    total_records: z.number(),
    total_pages: z.number(),
    current_page: z.number(),
    next_page: z.nullable(z.number()),
    prev_page: z.nullable(z.number()),
}).transform((v) => {
    return (0, primitives_js_1.remap)(v, {
        "total_records": "totalRecords",
        "total_pages": "totalPages",
        "current_page": "currentPage",
        "next_page": "nextPage",
        "prev_page": "prevPage",
    });
});
/** @internal */
exports.PaginationEntity$outboundSchema = z.object({
    totalRecords: z.number(),
    totalPages: z.number(),
    currentPage: z.number(),
    nextPage: z.nullable(z.number()),
    prevPage: z.nullable(z.number()),
}).transform((v) => {
    return (0, primitives_js_1.remap)(v, {
        totalRecords: "total_records",
        totalPages: "total_pages",
        currentPage: "current_page",
        nextPage: "next_page",
        prevPage: "prev_page",
    });
});
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
var PaginationEntity$;
(function (PaginationEntity$) {
    /** @deprecated use `PaginationEntity$inboundSchema` instead. */
    PaginationEntity$.inboundSchema = exports.PaginationEntity$inboundSchema;
    /** @deprecated use `PaginationEntity$outboundSchema` instead. */
    PaginationEntity$.outboundSchema = exports.PaginationEntity$outboundSchema;
})(PaginationEntity$ || (exports.PaginationEntity$ = PaginationEntity$ = {}));
function paginationEntityToJSON(paginationEntity) {
    return JSON.stringify(exports.PaginationEntity$outboundSchema.parse(paginationEntity));
}
function paginationEntityFromJSON(jsonString) {
    return (0, schemas_js_1.safeParse)(jsonString, (x) => exports.PaginationEntity$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'PaginationEntity' from JSON`);
}
