// creem/dist/commonjs/models/components/productentity.js
"use strict";
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProductEntity$ = exports.ProductEntity$outboundSchema = exports.ProductEntity$inboundSchema = exports.Mode$ = exports.Mode$outboundSchema = exports.Mode$inboundSchema = exports.Mode = void 0;
exports.productEntityToJSON = productEntityToJSON;
exports.productEntityFromJSON = productEntityFromJSON;
const z = __importStar(require("zod"));
const primitives_js_1 = require("../../lib/primitives.js");
const schemas_js_1 = require("../../lib/schemas.js");
const featureentity_js_1 = require("./featureentity.js");
/**
 * String representing the environment.
 */
exports.Mode = {
    Test: "test",
    Prod: "prod",
    Sandbox: "sandbox",
};
/** @internal */
exports.Mode$inboundSchema = z.nativeEnum(exports.Mode);
/** @internal */
exports.Mode$outboundSchema = exports.Mode$inboundSchema;
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
var Mode$;
(function (Mode$) {
    /** @deprecated use `Mode$inboundSchema` instead. */
    Mode$.inboundSchema = exports.Mode$inboundSchema;
    /** @deprecated use `Mode$outboundSchema` instead. */
    Mode$.outboundSchema = exports.Mode$outboundSchema;
})(Mode$ || (exports.Mode$ = Mode$ = {}));
/** @internal */
exports.ProductEntity$inboundSchema = z.object({
    id: z.string(),
    mode: exports.Mode$inboundSchema,
    object: z.string(),
    name: z.string(),
    description: z.string(),
    image_url: z.string().optional(),
    features: z.array(featureentity_js_1.FeatureEntity$inboundSchema).optional(),
    price: z.number(),
    currency: z.string(),
    billing_type: z.string(),
    billing_period: z.string(),
    status: z.string(),
    tax_mode: z.string(),
    tax_category: z.string(),
    product_url: z.string().optional(),
    default_success_url: z.nullable(z.string()).optional(),
    created_at: z.string().datetime({ offset: true }).transform(v => new Date(v)),
    updated_at: z.string().datetime({ offset: true }).transform(v => new Date(v)),
}).transform((v) => {
    return (0, primitives_js_1.remap)(v, {
        "image_url": "imageUrl",
        "billing_type": "billingType",
        "billing_period": "billingPeriod",
        "tax_mode": "taxMode",
        "tax_category": "taxCategory",
        "product_url": "productUrl",
        "default_success_url": "defaultSuccessUrl",
        "created_at": "createdAt",
        "updated_at": "updatedAt",
    });
});
/** @internal */
exports.ProductEntity$outboundSchema = z.object({
    id: z.string(),
    mode: exports.Mode$outboundSchema,
    object: z.string(),
    name: z.string(),
    description: z.string(),
    imageUrl: z.string().optional(),
    features: z.array(featureentity_js_1.FeatureEntity$outboundSchema).optional(),
    price: z.number(),
    currency: z.string(),
    billingType: z.string(),
    billingPeriod: z.string(),
    status: z.string(),
    taxMode: z.string(),
    taxCategory: z.string(),
    productUrl: z.string().optional(),
    defaultSuccessUrl: z.nullable(z.string()).optional(),
    createdAt: z.date().transform(v => v.toISOString()),
    updatedAt: z.date().transform(v => v.toISOString()),
}).transform((v) => {
    return (0, primitives_js_1.remap)(v, {
        imageUrl: "image_url",
        billingType: "billing_type",
        billingPeriod: "billing_period",
        taxMode: "tax_mode",
        taxCategory: "tax_category",
        productUrl: "product_url",
        defaultSuccessUrl: "default_success_url",
        createdAt: "created_at",
        updatedAt: "updated_at",
    });
});
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
var ProductEntity$;
(function (ProductEntity$) {
    /** @deprecated use `ProductEntity$inboundSchema` instead. */
    ProductEntity$.inboundSchema = exports.ProductEntity$inboundSchema;
    /** @deprecated use `ProductEntity$outboundSchema` instead. */
    ProductEntity$.outboundSchema = exports.ProductEntity$outboundSchema;
})(ProductEntity$ || (exports.ProductEntity$ = ProductEntity$ = {}));
function productEntityToJSON(productEntity) {
    return JSON.stringify(exports.ProductEntity$outboundSchema.parse(productEntity));
}
function productEntityFromJSON(jsonString) {
    return (0, schemas_js_1.safeParse)(jsonString, (x) => exports.ProductEntity$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ProductEntity' from JSON`);
}
