// creem/dist/commonjs/models/components/subscriptionitementity.js
"use strict";
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SubscriptionItemEntity$ = exports.SubscriptionItemEntity$outboundSchema = exports.SubscriptionItemEntity$inboundSchema = exports.SubscriptionItemEntityMode$ = exports.SubscriptionItemEntityMode$outboundSchema = exports.SubscriptionItemEntityMode$inboundSchema = exports.SubscriptionItemEntityMode = void 0;
exports.subscriptionItemEntityToJSON = subscriptionItemEntityToJSON;
exports.subscriptionItemEntityFromJSON = subscriptionItemEntityFromJSON;
const z = __importStar(require("zod"));
const primitives_js_1 = require("../../lib/primitives.js");
const schemas_js_1 = require("../../lib/schemas.js");
/**
 * String representing the environment.
 */
exports.SubscriptionItemEntityMode = {
    Test: "test",
    Prod: "prod",
    Sandbox: "sandbox",
};
/** @internal */
exports.SubscriptionItemEntityMode$inboundSchema = z.nativeEnum(exports.SubscriptionItemEntityMode);
/** @internal */
exports.SubscriptionItemEntityMode$outboundSchema = exports.SubscriptionItemEntityMode$inboundSchema;
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
var SubscriptionItemEntityMode$;
(function (SubscriptionItemEntityMode$) {
    /** @deprecated use `SubscriptionItemEntityMode$inboundSchema` instead. */
    SubscriptionItemEntityMode$.inboundSchema = exports.SubscriptionItemEntityMode$inboundSchema;
    /** @deprecated use `SubscriptionItemEntityMode$outboundSchema` instead. */
    SubscriptionItemEntityMode$.outboundSchema = exports.SubscriptionItemEntityMode$outboundSchema;
})(SubscriptionItemEntityMode$ || (exports.SubscriptionItemEntityMode$ = SubscriptionItemEntityMode$ = {}));
/** @internal */
exports.SubscriptionItemEntity$inboundSchema = z.object({
    id: z.string(),
    mode: exports.SubscriptionItemEntityMode$inboundSchema,
    object: z.string(),
    product_id: z.string().optional(),
    price_id: z.string().optional(),
    units: z.number().optional(),
}).transform((v) => {
    return (0, primitives_js_1.remap)(v, {
        "product_id": "productId",
        "price_id": "priceId",
    });
});
/** @internal */
exports.SubscriptionItemEntity$outboundSchema = z.object({
    id: z.string(),
    mode: exports.SubscriptionItemEntityMode$outboundSchema,
    object: z.string(),
    productId: z.string().optional(),
    priceId: z.string().optional(),
    units: z.number().optional(),
}).transform((v) => {
    return (0, primitives_js_1.remap)(v, {
        productId: "product_id",
        priceId: "price_id",
    });
});
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
var SubscriptionItemEntity$;
(function (SubscriptionItemEntity$) {
    /** @deprecated use `SubscriptionItemEntity$inboundSchema` instead. */
    SubscriptionItemEntity$.inboundSchema = exports.SubscriptionItemEntity$inboundSchema;
    /** @deprecated use `SubscriptionItemEntity$outboundSchema` instead. */
    SubscriptionItemEntity$.outboundSchema = exports.SubscriptionItemEntity$outboundSchema;
})(SubscriptionItemEntity$ || (exports.SubscriptionItemEntity$ = SubscriptionItemEntity$ = {}));
function subscriptionItemEntityToJSON(subscriptionItemEntity) {
    return JSON.stringify(exports.SubscriptionItemEntity$outboundSchema.parse(subscriptionItemEntity));
}
function subscriptionItemEntityFromJSON(jsonString) {
    return (0, schemas_js_1.safeParse)(jsonString, (x) => exports.SubscriptionItemEntity$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'SubscriptionItemEntity' from JSON`);
}
