// creem/dist/commonjs/models/components/transactionlistentity.js
"use strict";
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionListEntity$ = exports.TransactionListEntity$outboundSchema = exports.TransactionListEntity$inboundSchema = void 0;
exports.transactionListEntityToJSON = transactionListEntityToJSON;
exports.transactionListEntityFromJSON = transactionListEntityFromJSON;
const z = __importStar(require("zod"));
const schemas_js_1 = require("../../lib/schemas.js");
const paginationentity_js_1 = require("./paginationentity.js");
const transactionentity_js_1 = require("./transactionentity.js");
/** @internal */
exports.TransactionListEntity$inboundSchema = z.object({
    items: z.array(transactionentity_js_1.TransactionEntity$inboundSchema),
    pagination: paginationentity_js_1.PaginationEntity$inboundSchema,
});
/** @internal */
exports.TransactionListEntity$outboundSchema = z.object({
    items: z.array(transactionentity_js_1.TransactionEntity$outboundSchema),
    pagination: paginationentity_js_1.PaginationEntity$outboundSchema,
});
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
var TransactionListEntity$;
(function (TransactionListEntity$) {
    /** @deprecated use `TransactionListEntity$inboundSchema` instead. */
    TransactionListEntity$.inboundSchema = exports.TransactionListEntity$inboundSchema;
    /** @deprecated use `TransactionListEntity$outboundSchema` instead. */
    TransactionListEntity$.outboundSchema = exports.TransactionListEntity$outboundSchema;
})(TransactionListEntity$ || (exports.TransactionListEntity$ = TransactionListEntity$ = {}));
function transactionListEntityToJSON(transactionListEntity) {
    return JSON.stringify(exports.TransactionListEntity$outboundSchema.parse(transactionListEntity));
}
function transactionListEntityFromJSON(jsonString) {
    return (0, schemas_js_1.safeParse)(jsonString, (x) => exports.TransactionListEntity$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'TransactionListEntity' from JSON`);
}
