// creem/dist/commonjs/models/components/updatesubscriptionrequestentity.js
"use strict";
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.UpdateSubscriptionRequestEntity$ = exports.UpdateSubscriptionRequestEntity$outboundSchema = exports.UpdateSubscriptionRequestEntity$inboundSchema = exports.UpdateBehavior$ = exports.UpdateBehavior$outboundSchema = exports.UpdateBehavior$inboundSchema = exports.UpdateBehavior = void 0;
exports.updateSubscriptionRequestEntityToJSON = updateSubscriptionRequestEntityToJSON;
exports.updateSubscriptionRequestEntityFromJSON = updateSubscriptionRequestEntityFromJSON;
const z = __importStar(require("zod"));
const primitives_js_1 = require("../../lib/primitives.js");
const schemas_js_1 = require("../../lib/schemas.js");
const upsertsubscriptionitementity_js_1 = require("./upsertsubscriptionitementity.js");
/**
 * The update behavior for the subscription (defaults to proration)
 */
exports.UpdateBehavior = {
    ProrationChargeImmediately: "proration-charge-immediately",
    ProrationCharge: "proration-charge",
    ProrationNone: "proration-none",
};
/** @internal */
exports.UpdateBehavior$inboundSchema = z.nativeEnum(exports.UpdateBehavior);
/** @internal */
exports.UpdateBehavior$outboundSchema = exports.UpdateBehavior$inboundSchema;
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
var UpdateBehavior$;
(function (UpdateBehavior$) {
    /** @deprecated use `UpdateBehavior$inboundSchema` instead. */
    UpdateBehavior$.inboundSchema = exports.UpdateBehavior$inboundSchema;
    /** @deprecated use `UpdateBehavior$outboundSchema` instead. */
    UpdateBehavior$.outboundSchema = exports.UpdateBehavior$outboundSchema;
})(UpdateBehavior$ || (exports.UpdateBehavior$ = UpdateBehavior$ = {}));
/** @internal */
exports.UpdateSubscriptionRequestEntity$inboundSchema = z.object({
    items: z.array(upsertsubscriptionitementity_js_1.UpsertSubscriptionItemEntity$inboundSchema).optional(),
    update_behavior: exports.UpdateBehavior$inboundSchema.default("proration-charge"),
}).transform((v) => {
    return (0, primitives_js_1.remap)(v, {
        "update_behavior": "updateBehavior",
    });
});
/** @internal */
exports.UpdateSubscriptionRequestEntity$outboundSchema = z.object({
    items: z.array(upsertsubscriptionitementity_js_1.UpsertSubscriptionItemEntity$outboundSchema).optional(),
    updateBehavior: exports.UpdateBehavior$outboundSchema.default("proration-charge"),
}).transform((v) => {
    return (0, primitives_js_1.remap)(v, {
        updateBehavior: "update_behavior",
    });
});
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
var UpdateSubscriptionRequestEntity$;
(function (UpdateSubscriptionRequestEntity$) {
    /** @deprecated use `UpdateSubscriptionRequestEntity$inboundSchema` instead. */
    UpdateSubscriptionRequestEntity$.inboundSchema = exports.UpdateSubscriptionRequestEntity$inboundSchema;
    /** @deprecated use `UpdateSubscriptionRequestEntity$outboundSchema` instead. */
    UpdateSubscriptionRequestEntity$.outboundSchema = exports.UpdateSubscriptionRequestEntity$outboundSchema;
})(UpdateSubscriptionRequestEntity$ || (exports.UpdateSubscriptionRequestEntity$ = UpdateSubscriptionRequestEntity$ = {}));
function updateSubscriptionRequestEntityToJSON(updateSubscriptionRequestEntity) {
    return JSON.stringify(exports.UpdateSubscriptionRequestEntity$outboundSchema.parse(updateSubscriptionRequestEntity));
}
function updateSubscriptionRequestEntityFromJSON(jsonString) {
    return (0, schemas_js_1.safeParse)(jsonString, (x) => exports.UpdateSubscriptionRequestEntity$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'UpdateSubscriptionRequestEntity' from JSON`);
}
