// creem/dist/esm/models/components/customerlinksentity.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
import * as z from "zod";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
/** @internal */
export const CustomerLinksEntity$inboundSchema = z.object({
    customer_portal_link: z.string(),
}).transform((v) => {
    return remap$(v, {
        "customer_portal_link": "customerPortalLink",
    });
});
/** @internal */
export const CustomerLinksEntity$outboundSchema = z.object({
    customerPortalLink: z.string(),
}).transform((v) => {
    return remap$(v, {
        customerPortalLink: "customer_portal_link",
    });
});
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
export var CustomerLinksEntity$;
(function (CustomerLinksEntity$) {
    /** @deprecated use `CustomerLinksEntity$inboundSchema` instead. */
    CustomerLinksEntity$.inboundSchema = CustomerLinksEntity$inboundSchema;
    /** @deprecated use `CustomerLinksEntity$outboundSchema` instead. */
    CustomerLinksEntity$.outboundSchema = CustomerLinksEntity$outboundSchema;
})(CustomerLinksEntity$ || (CustomerLinksEntity$ = {}));
export function customerLinksEntityToJSON(customerLinksEntity) {
    return JSON.stringify(CustomerLinksEntity$outboundSchema.parse(customerLinksEntity));
}
export function customerLinksEntityFromJSON(jsonString) {
    return safeParse(jsonString, (x) => CustomerLinksEntity$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'CustomerLinksEntity' from JSON`);
}
