// creem/dist/esm/models/components/paginationentity.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
import * as z from "zod";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
/** @internal */
export const PaginationEntity$inboundSchema = z.object({
    total_records: z.number(),
    total_pages: z.number(),
    current_page: z.number(),
    next_page: z.nullable(z.number()),
    prev_page: z.nullable(z.number()),
}).transform((v) => {
    return remap$(v, {
        "total_records": "totalRecords",
        "total_pages": "totalPages",
        "current_page": "currentPage",
        "next_page": "nextPage",
        "prev_page": "prevPage",
    });
});
/** @internal */
export const PaginationEntity$outboundSchema = z.object({
    totalRecords: z.number(),
    totalPages: z.number(),
    currentPage: z.number(),
    nextPage: z.nullable(z.number()),
    prevPage: z.nullable(z.number()),
}).transform((v) => {
    return remap$(v, {
        totalRecords: "total_records",
        totalPages: "total_pages",
        currentPage: "current_page",
        nextPage: "next_page",
        prevPage: "prev_page",
    });
});
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
export var PaginationEntity$;
(function (PaginationEntity$) {
    /** @deprecated use `PaginationEntity$inboundSchema` instead. */
    PaginationEntity$.inboundSchema = PaginationEntity$inboundSchema;
    /** @deprecated use `PaginationEntity$outboundSchema` instead. */
    PaginationEntity$.outboundSchema = PaginationEntity$outboundSchema;
})(PaginationEntity$ || (PaginationEntity$ = {}));
export function paginationEntityToJSON(paginationEntity) {
    return JSON.stringify(PaginationEntity$outboundSchema.parse(paginationEntity));
}
export function paginationEntityFromJSON(jsonString) {
    return safeParse(jsonString, (x) => PaginationEntity$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'PaginationEntity' from JSON`);
}
