// creem/dist/esm/models/components/transactionlistentity.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
import * as z from "zod";
import { safeParse } from "../../lib/schemas.js";
import { PaginationEntity$inboundSchema, PaginationEntity$outboundSchema, } from "./paginationentity.js";
import { TransactionEntity$inboundSchema, TransactionEntity$outboundSchema, } from "./transactionentity.js";
/** @internal */
export const TransactionListEntity$inboundSchema = z.object({
    items: z.array(TransactionEntity$inboundSchema),
    pagination: PaginationEntity$inboundSchema,
});
/** @internal */
export const TransactionListEntity$outboundSchema = z.object({
    items: z.array(TransactionEntity$outboundSchema),
    pagination: PaginationEntity$outboundSchema,
});
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
export var TransactionListEntity$;
(function (TransactionListEntity$) {
    /** @deprecated use `TransactionListEntity$inboundSchema` instead. */
    TransactionListEntity$.inboundSchema = TransactionListEntity$inboundSchema;
    /** @deprecated use `TransactionListEntity$outboundSchema` instead. */
    TransactionListEntity$.outboundSchema = TransactionListEntity$outboundSchema;
})(TransactionListEntity$ || (TransactionListEntity$ = {}));
export function transactionListEntityToJSON(transactionListEntity) {
    return JSON.stringify(TransactionListEntity$outboundSchema.parse(transactionListEntity));
}
export function transactionListEntityFromJSON(jsonString) {
    return safeParse(jsonString, (x) => TransactionListEntity$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'TransactionListEntity' from JSON`);
}
