// creem/dist/commonjs/models/components/createproductrequestentity.js
"use strict";
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateProductRequestEntity$ = exports.CreateProductRequestEntity$outboundSchema = exports.CreateProductRequestEntity$inboundSchema = void 0;
exports.createProductRequestEntityToJSON = createProductRequestEntityToJSON;
exports.createProductRequestEntityFromJSON = createProductRequestEntityFromJSON;
const z = __importStar(require("zod"));
const primitives_js_1 = require("../../lib/primitives.js");
const schemas_js_1 = require("../../lib/schemas.js");
const customfieldrequestentity_js_1 = require("./customfieldrequestentity.js");
/** @internal */
exports.CreateProductRequestEntity$inboundSchema = z.object({
    name: z.string(),
    description: z.string().optional(),
    image_url: z.string().optional(),
    price: z.number().int(),
    currency: z.string(),
    billing_type: z.string(),
    billing_period: z.string().optional(),
    tax_mode: z.string().optional(),
    tax_category: z.string().optional(),
    default_success_url: z.string().optional(),
    custom_field: z.array(customfieldrequestentity_js_1.CustomFieldRequestEntity$inboundSchema).optional(),
}).transform((v) => {
    return (0, primitives_js_1.remap)(v, {
        "image_url": "imageUrl",
        "billing_type": "billingType",
        "billing_period": "billingPeriod",
        "tax_mode": "taxMode",
        "tax_category": "taxCategory",
        "default_success_url": "defaultSuccessUrl",
        "custom_field": "customField",
    });
});
/** @internal */
exports.CreateProductRequestEntity$outboundSchema = z.object({
    name: z.string(),
    description: z.string().optional(),
    imageUrl: z.string().optional(),
    price: z.number().int(),
    currency: z.string(),
    billingType: z.string(),
    billingPeriod: z.string().optional(),
    taxMode: z.string().optional(),
    taxCategory: z.string().optional(),
    defaultSuccessUrl: z.string().optional(),
    customField: z.array(customfieldrequestentity_js_1.CustomFieldRequestEntity$outboundSchema).optional(),
}).transform((v) => {
    return (0, primitives_js_1.remap)(v, {
        imageUrl: "image_url",
        billingType: "billing_type",
        billingPeriod: "billing_period",
        taxMode: "tax_mode",
        taxCategory: "tax_category",
        defaultSuccessUrl: "default_success_url",
        customField: "custom_field",
    });
});
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
var CreateProductRequestEntity$;
(function (CreateProductRequestEntity$) {
    /** @deprecated use `CreateProductRequestEntity$inboundSchema` instead. */
    CreateProductRequestEntity$.inboundSchema = exports.CreateProductRequestEntity$inboundSchema;
    /** @deprecated use `CreateProductRequestEntity$outboundSchema` instead. */
    CreateProductRequestEntity$.outboundSchema = exports.CreateProductRequestEntity$outboundSchema;
})(CreateProductRequestEntity$ || (exports.CreateProductRequestEntity$ = CreateProductRequestEntity$ = {}));
function createProductRequestEntityToJSON(createProductRequestEntity) {
    return JSON.stringify(exports.CreateProductRequestEntity$outboundSchema.parse(createProductRequestEntity));
}
function createProductRequestEntityFromJSON(jsonString) {
    return (0, schemas_js_1.safeParse)(jsonString, (x) => exports.CreateProductRequestEntity$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'CreateProductRequestEntity' from JSON`);
}
