// creem/dist/commonjs/models/components/customfieldrequestentity.js
"use strict";
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomFieldRequestEntity$ = exports.CustomFieldRequestEntity$outboundSchema = exports.CustomFieldRequestEntity$inboundSchema = exports.Type$ = exports.Type$outboundSchema = exports.Type$inboundSchema = exports.Type = void 0;
exports.customFieldRequestEntityToJSON = customFieldRequestEntityToJSON;
exports.customFieldRequestEntityFromJSON = customFieldRequestEntityFromJSON;
const z = __importStar(require("zod"));
const schemas_js_1 = require("../../lib/schemas.js");
const text_js_1 = require("./text.js");
/**
 * The type of the field.
 */
exports.Type = {
    Text: "text",
};
/** @internal */
exports.Type$inboundSchema = z.nativeEnum(exports.Type);
/** @internal */
exports.Type$outboundSchema = exports.Type$inboundSchema;
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
var Type$;
(function (Type$) {
    /** @deprecated use `Type$inboundSchema` instead. */
    Type$.inboundSchema = exports.Type$inboundSchema;
    /** @deprecated use `Type$outboundSchema` instead. */
    Type$.outboundSchema = exports.Type$outboundSchema;
})(Type$ || (exports.Type$ = Type$ = {}));
/** @internal */
exports.CustomFieldRequestEntity$inboundSchema = z.object({
    type: exports.Type$inboundSchema,
    key: z.string(),
    label: z.string(),
    optional: z.boolean().optional(),
    text: text_js_1.Text$inboundSchema.optional(),
});
/** @internal */
exports.CustomFieldRequestEntity$outboundSchema = z.object({
    type: exports.Type$outboundSchema,
    key: z.string(),
    label: z.string(),
    optional: z.boolean().optional(),
    text: text_js_1.Text$outboundSchema.optional(),
});
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
var CustomFieldRequestEntity$;
(function (CustomFieldRequestEntity$) {
    /** @deprecated use `CustomFieldRequestEntity$inboundSchema` instead. */
    CustomFieldRequestEntity$.inboundSchema = exports.CustomFieldRequestEntity$inboundSchema;
    /** @deprecated use `CustomFieldRequestEntity$outboundSchema` instead. */
    CustomFieldRequestEntity$.outboundSchema = exports.CustomFieldRequestEntity$outboundSchema;
})(CustomFieldRequestEntity$ || (exports.CustomFieldRequestEntity$ = CustomFieldRequestEntity$ = {}));
function customFieldRequestEntityToJSON(customFieldRequestEntity) {
    return JSON.stringify(exports.CustomFieldRequestEntity$outboundSchema.parse(customFieldRequestEntity));
}
function customFieldRequestEntityFromJSON(jsonString) {
    return (0, schemas_js_1.safeParse)(jsonString, (x) => exports.CustomFieldRequestEntity$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'CustomFieldRequestEntity' from JSON`);
}
