// creem/dist/commonjs/models/components/upgradesubscriptionrequestentity.js
"use strict";
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.UpgradeSubscriptionRequestEntity$ = exports.UpgradeSubscriptionRequestEntity$outboundSchema = exports.UpgradeSubscriptionRequestEntity$inboundSchema = exports.UpgradeSubscriptionRequestEntityUpdateBehavior$ = exports.UpgradeSubscriptionRequestEntityUpdateBehavior$outboundSchema = exports.UpgradeSubscriptionRequestEntityUpdateBehavior$inboundSchema = exports.UpgradeSubscriptionRequestEntityUpdateBehavior = void 0;
exports.upgradeSubscriptionRequestEntityToJSON = upgradeSubscriptionRequestEntityToJSON;
exports.upgradeSubscriptionRequestEntityFromJSON = upgradeSubscriptionRequestEntityFromJSON;
const z = __importStar(require("zod"));
const primitives_js_1 = require("../../lib/primitives.js");
const schemas_js_1 = require("../../lib/schemas.js");
/**
 * The update behavior for the subscription (defaults to proration-charge-immediately)
 */
exports.UpgradeSubscriptionRequestEntityUpdateBehavior = {
    ProrationChargeImmediately: "proration-charge-immediately",
    ProrationCharge: "proration-charge",
    ProrationNone: "proration-none",
};
/** @internal */
exports.UpgradeSubscriptionRequestEntityUpdateBehavior$inboundSchema = z
    .nativeEnum(exports.UpgradeSubscriptionRequestEntityUpdateBehavior);
/** @internal */
exports.UpgradeSubscriptionRequestEntityUpdateBehavior$outboundSchema = exports.UpgradeSubscriptionRequestEntityUpdateBehavior$inboundSchema;
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
var UpgradeSubscriptionRequestEntityUpdateBehavior$;
(function (UpgradeSubscriptionRequestEntityUpdateBehavior$) {
    /** @deprecated use `UpgradeSubscriptionRequestEntityUpdateBehavior$inboundSchema` instead. */
    UpgradeSubscriptionRequestEntityUpdateBehavior$.inboundSchema = exports.UpgradeSubscriptionRequestEntityUpdateBehavior$inboundSchema;
    /** @deprecated use `UpgradeSubscriptionRequestEntityUpdateBehavior$outboundSchema` instead. */
    UpgradeSubscriptionRequestEntityUpdateBehavior$.outboundSchema = exports.UpgradeSubscriptionRequestEntityUpdateBehavior$outboundSchema;
})(UpgradeSubscriptionRequestEntityUpdateBehavior$ || (exports.UpgradeSubscriptionRequestEntityUpdateBehavior$ = UpgradeSubscriptionRequestEntityUpdateBehavior$ = {}));
/** @internal */
exports.UpgradeSubscriptionRequestEntity$inboundSchema = z.object({
    product_id: z.string(),
    update_behavior: exports.UpgradeSubscriptionRequestEntityUpdateBehavior$inboundSchema
        .default("proration-charge-immediately"),
}).transform((v) => {
    return (0, primitives_js_1.remap)(v, {
        "product_id": "productId",
        "update_behavior": "updateBehavior",
    });
});
/** @internal */
exports.UpgradeSubscriptionRequestEntity$outboundSchema = z.object({
    productId: z.string(),
    updateBehavior: exports.UpgradeSubscriptionRequestEntityUpdateBehavior$outboundSchema
        .default("proration-charge-immediately"),
}).transform((v) => {
    return (0, primitives_js_1.remap)(v, {
        productId: "product_id",
        updateBehavior: "update_behavior",
    });
});
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
var UpgradeSubscriptionRequestEntity$;
(function (UpgradeSubscriptionRequestEntity$) {
    /** @deprecated use `UpgradeSubscriptionRequestEntity$inboundSchema` instead. */
    UpgradeSubscriptionRequestEntity$.inboundSchema = exports.UpgradeSubscriptionRequestEntity$inboundSchema;
    /** @deprecated use `UpgradeSubscriptionRequestEntity$outboundSchema` instead. */
    UpgradeSubscriptionRequestEntity$.outboundSchema = exports.UpgradeSubscriptionRequestEntity$outboundSchema;
})(UpgradeSubscriptionRequestEntity$ || (exports.UpgradeSubscriptionRequestEntity$ = UpgradeSubscriptionRequestEntity$ = {}));
function upgradeSubscriptionRequestEntityToJSON(upgradeSubscriptionRequestEntity) {
    return JSON.stringify(exports.UpgradeSubscriptionRequestEntity$outboundSchema.parse(upgradeSubscriptionRequestEntity));
}
function upgradeSubscriptionRequestEntityFromJSON(jsonString) {
    return (0, schemas_js_1.safeParse)(jsonString, (x) => exports.UpgradeSubscriptionRequestEntity$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'UpgradeSubscriptionRequestEntity' from JSON`);
}
