// creem/dist/esm/models/operations/deactivatelicense.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
import * as z from "zod";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import * as components from "../components/index.js";
/** @internal */
export const DeactivateLicenseRequest$inboundSchema = z.object({
    "x-api-key": z.string(),
    DeactivateLicenseRequestEntity: components.DeactivateLicenseRequestEntity$inboundSchema,
}).transform((v) => {
    return remap$(v, {
        "x-api-key": "xApiKey",
        "DeactivateLicenseRequestEntity": "deactivateLicenseRequestEntity",
    });
});
/** @internal */
export const DeactivateLicenseRequest$outboundSchema = z.object({
    xApiKey: z.string(),
    deactivateLicenseRequestEntity: components.DeactivateLicenseRequestEntity$outboundSchema,
}).transform((v) => {
    return remap$(v, {
        xApiKey: "x-api-key",
        deactivateLicenseRequestEntity: "DeactivateLicenseRequestEntity",
    });
});
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
export var DeactivateLicenseRequest$;
(function (DeactivateLicenseRequest$) {
    /** @deprecated use `DeactivateLicenseRequest$inboundSchema` instead. */
    DeactivateLicenseRequest$.inboundSchema = DeactivateLicenseRequest$inboundSchema;
    /** @deprecated use `DeactivateLicenseRequest$outboundSchema` instead. */
    DeactivateLicenseRequest$.outboundSchema = DeactivateLicenseRequest$outboundSchema;
})(DeactivateLicenseRequest$ || (DeactivateLicenseRequest$ = {}));
export function deactivateLicenseRequestToJSON(deactivateLicenseRequest) {
    return JSON.stringify(DeactivateLicenseRequest$outboundSchema.parse(deactivateLicenseRequest));
}
export function deactivateLicenseRequestFromJSON(jsonString) {
    return safeParse(jsonString, (x) => DeactivateLicenseRequest$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'DeactivateLicenseRequest' from JSON`);
}
