// creem/dist/esm/models/operations/deletediscount.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
import * as z from "zod";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
/** @internal */
export const DeleteDiscountRequest$inboundSchema = z.object({
    id: z.string(),
    "x-api-key": z.string(),
}).transform((v) => {
    return remap$(v, {
        "x-api-key": "xApiKey",
    });
});
/** @internal */
export const DeleteDiscountRequest$outboundSchema = z.object({
    id: z.string(),
    xApiKey: z.string(),
}).transform((v) => {
    return remap$(v, {
        xApiKey: "x-api-key",
    });
});
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
export var DeleteDiscountRequest$;
(function (DeleteDiscountRequest$) {
    /** @deprecated use `DeleteDiscountRequest$inboundSchema` instead. */
    DeleteDiscountRequest$.inboundSchema = DeleteDiscountRequest$inboundSchema;
    /** @deprecated use `DeleteDiscountRequest$outboundSchema` instead. */
    DeleteDiscountRequest$.outboundSchema = DeleteDiscountRequest$outboundSchema;
})(DeleteDiscountRequest$ || (DeleteDiscountRequest$ = {}));
export function deleteDiscountRequestToJSON(deleteDiscountRequest) {
    return JSON.stringify(DeleteDiscountRequest$outboundSchema.parse(deleteDiscountRequest));
}
export function deleteDiscountRequestFromJSON(jsonString) {
    return safeParse(jsonString, (x) => DeleteDiscountRequest$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'DeleteDiscountRequest' from JSON`);
}
