// creem/dist/esm/models/operations/generatecustomerlinks.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
import * as z from "zod";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import * as components from "../components/index.js";
/** @internal */
export const GenerateCustomerLinksRequest$inboundSchema = z.object({
    "x-api-key": z.string(),
    CreateCustomerPortalLinkRequestEntity: components.CreateCustomerPortalLinkRequestEntity$inboundSchema,
}).transform((v) => {
    return remap$(v, {
        "x-api-key": "xApiKey",
        "CreateCustomerPortalLinkRequestEntity": "createCustomerPortalLinkRequestEntity",
    });
});
/** @internal */
export const GenerateCustomerLinksRequest$outboundSchema = z.object({
    xApiKey: z.string(),
    createCustomerPortalLinkRequestEntity: components.CreateCustomerPortalLinkRequestEntity$outboundSchema,
}).transform((v) => {
    return remap$(v, {
        xApiKey: "x-api-key",
        createCustomerPortalLinkRequestEntity: "CreateCustomerPortalLinkRequestEntity",
    });
});
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
export var GenerateCustomerLinksRequest$;
(function (GenerateCustomerLinksRequest$) {
    /** @deprecated use `GenerateCustomerLinksRequest$inboundSchema` instead. */
    GenerateCustomerLinksRequest$.inboundSchema = GenerateCustomerLinksRequest$inboundSchema;
    /** @deprecated use `GenerateCustomerLinksRequest$outboundSchema` instead. */
    GenerateCustomerLinksRequest$.outboundSchema = GenerateCustomerLinksRequest$outboundSchema;
})(GenerateCustomerLinksRequest$ || (GenerateCustomerLinksRequest$ = {}));
export function generateCustomerLinksRequestToJSON(generateCustomerLinksRequest) {
    return JSON.stringify(GenerateCustomerLinksRequest$outboundSchema.parse(generateCustomerLinksRequest));
}
export function generateCustomerLinksRequestFromJSON(jsonString) {
    return safeParse(jsonString, (x) => GenerateCustomerLinksRequest$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'GenerateCustomerLinksRequest' from JSON`);
}
