// creem/dist/esm/models/operations/upgradesubscription.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
import * as z from "zod";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import * as components from "../components/index.js";
/** @internal */
export const UpgradeSubscriptionRequest$inboundSchema = z.object({
    id: z.string(),
    "x-api-key": z.string(),
    UpgradeSubscriptionRequestEntity: components.UpgradeSubscriptionRequestEntity$inboundSchema,
}).transform((v) => {
    return remap$(v, {
        "x-api-key": "xApiKey",
        "UpgradeSubscriptionRequestEntity": "upgradeSubscriptionRequestEntity",
    });
});
/** @internal */
export const UpgradeSubscriptionRequest$outboundSchema = z.object({
    id: z.string(),
    xApiKey: z.string(),
    upgradeSubscriptionRequestEntity: components.UpgradeSubscriptionRequestEntity$outboundSchema,
}).transform((v) => {
    return remap$(v, {
        xApiKey: "x-api-key",
        upgradeSubscriptionRequestEntity: "UpgradeSubscriptionRequestEntity",
    });
});
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
export var UpgradeSubscriptionRequest$;
(function (UpgradeSubscriptionRequest$) {
    /** @deprecated use `UpgradeSubscriptionRequest$inboundSchema` instead. */
    UpgradeSubscriptionRequest$.inboundSchema = UpgradeSubscriptionRequest$inboundSchema;
    /** @deprecated use `UpgradeSubscriptionRequest$outboundSchema` instead. */
    UpgradeSubscriptionRequest$.outboundSchema = UpgradeSubscriptionRequest$outboundSchema;
})(UpgradeSubscriptionRequest$ || (UpgradeSubscriptionRequest$ = {}));
export function upgradeSubscriptionRequestToJSON(upgradeSubscriptionRequest) {
    return JSON.stringify(UpgradeSubscriptionRequest$outboundSchema.parse(upgradeSubscriptionRequest));
}
export function upgradeSubscriptionRequestFromJSON(jsonString) {
    return safeParse(jsonString, (x) => UpgradeSubscriptionRequest$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'UpgradeSubscriptionRequest' from JSON`);
}
