// creem/dist/esm/models/operations/creatediscount.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
import * as z from "zod";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import * as components from "../components/index.js";
/** @internal */
export const CreateDiscountRequest$inboundSchema = z.object({
    "x-api-key": z.string(),
    CreateDiscountRequestEntity: components.CreateDiscountRequestEntity$inboundSchema,
}).transform((v) => {
    return remap$(v, {
        "x-api-key": "xApiKey",
        "CreateDiscountRequestEntity": "createDiscountRequestEntity",
    });
});
/** @internal */
export const CreateDiscountRequest$outboundSchema = z.object({
    xApiKey: z.string(),
    createDiscountRequestEntity: components.CreateDiscountRequestEntity$outboundSchema,
}).transform((v) => {
    return remap$(v, {
        xApiKey: "x-api-key",
        createDiscountRequestEntity: "CreateDiscountRequestEntity",
    });
});
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
export var CreateDiscountRequest$;
(function (CreateDiscountRequest$) {
    /** @deprecated use `CreateDiscountRequest$inboundSchema` instead. */
    CreateDiscountRequest$.inboundSchema = CreateDiscountRequest$inboundSchema;
    /** @deprecated use `CreateDiscountRequest$outboundSchema` instead. */
    CreateDiscountRequest$.outboundSchema = CreateDiscountRequest$outboundSchema;
})(CreateDiscountRequest$ || (CreateDiscountRequest$ = {}));
export function createDiscountRequestToJSON(createDiscountRequest) {
    return JSON.stringify(CreateDiscountRequest$outboundSchema.parse(createDiscountRequest));
}
export function createDiscountRequestFromJSON(jsonString) {
    return safeParse(jsonString, (x) => CreateDiscountRequest$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'CreateDiscountRequest' from JSON`);
}
