// creem/dist/esm/models/operations/searchproducts.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
import * as z from "zod";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
/** @internal */
export const SearchProductsRequest$inboundSchema = z.object({
    page_number: z.number().optional(),
    page_size: z.number().optional(),
    "x-api-key": z.string(),
}).transform((v) => {
    return remap$(v, {
        "page_number": "pageNumber",
        "page_size": "pageSize",
        "x-api-key": "xApiKey",
    });
});
/** @internal */
export const SearchProductsRequest$outboundSchema = z.object({
    pageNumber: z.number().optional(),
    pageSize: z.number().optional(),
    xApiKey: z.string(),
}).transform((v) => {
    return remap$(v, {
        pageNumber: "page_number",
        pageSize: "page_size",
        xApiKey: "x-api-key",
    });
});
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
export var SearchProductsRequest$;
(function (SearchProductsRequest$) {
    /** @deprecated use `SearchProductsRequest$inboundSchema` instead. */
    SearchProductsRequest$.inboundSchema = SearchProductsRequest$inboundSchema;
    /** @deprecated use `SearchProductsRequest$outboundSchema` instead. */
    SearchProductsRequest$.outboundSchema = SearchProductsRequest$outboundSchema;
})(SearchProductsRequest$ || (SearchProductsRequest$ = {}));
export function searchProductsRequestToJSON(searchProductsRequest) {
    return JSON.stringify(SearchProductsRequest$outboundSchema.parse(searchProductsRequest));
}
export function searchProductsRequestFromJSON(jsonString) {
    return safeParse(jsonString, (x) => SearchProductsRequest$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'SearchProductsRequest' from JSON`);
}
