// creem/dist/esm/models/operations/updatesubscription.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
import * as z from "zod";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import * as components from "../components/index.js";
/** @internal */
export const UpdateSubscriptionRequest$inboundSchema = z.object({
    id: z.string(),
    "x-api-key": z.string(),
    UpdateSubscriptionRequestEntity: components.UpdateSubscriptionRequestEntity$inboundSchema,
}).transform((v) => {
    return remap$(v, {
        "x-api-key": "xApiKey",
        "UpdateSubscriptionRequestEntity": "updateSubscriptionRequestEntity",
    });
});
/** @internal */
export const UpdateSubscriptionRequest$outboundSchema = z.object({
    id: z.string(),
    xApiKey: z.string(),
    updateSubscriptionRequestEntity: components.UpdateSubscriptionRequestEntity$outboundSchema,
}).transform((v) => {
    return remap$(v, {
        xApiKey: "x-api-key",
        updateSubscriptionRequestEntity: "UpdateSubscriptionRequestEntity",
    });
});
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
export var UpdateSubscriptionRequest$;
(function (UpdateSubscriptionRequest$) {
    /** @deprecated use `UpdateSubscriptionRequest$inboundSchema` instead. */
    UpdateSubscriptionRequest$.inboundSchema = UpdateSubscriptionRequest$inboundSchema;
    /** @deprecated use `UpdateSubscriptionRequest$outboundSchema` instead. */
    UpdateSubscriptionRequest$.outboundSchema = UpdateSubscriptionRequest$outboundSchema;
})(UpdateSubscriptionRequest$ || (UpdateSubscriptionRequest$ = {}));
export function updateSubscriptionRequestToJSON(updateSubscriptionRequest) {
    return JSON.stringify(UpdateSubscriptionRequest$outboundSchema.parse(updateSubscriptionRequest));
}
export function updateSubscriptionRequestFromJSON(jsonString) {
    return safeParse(jsonString, (x) => UpdateSubscriptionRequest$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'UpdateSubscriptionRequest' from JSON`);
}
