// creem/dist/esm/models/components/deactivatelicenserequestentity.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
import * as z from "zod";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
/** @internal */
export const DeactivateLicenseRequestEntity$inboundSchema = z.object({
    key: z.string(),
    instance_id: z.string(),
}).transform((v) => {
    return remap$(v, {
        "instance_id": "instanceId",
    });
});
/** @internal */
export const DeactivateLicenseRequestEntity$outboundSchema = z.object({
    key: z.string(),
    instanceId: z.string(),
}).transform((v) => {
    return remap$(v, {
        instanceId: "instance_id",
    });
});
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
export var DeactivateLicenseRequestEntity$;
(function (DeactivateLicenseRequestEntity$) {
    /** @deprecated use `DeactivateLicenseRequestEntity$inboundSchema` instead. */
    DeactivateLicenseRequestEntity$.inboundSchema = DeactivateLicenseRequestEntity$inboundSchema;
    /** @deprecated use `DeactivateLicenseRequestEntity$outboundSchema` instead. */
    DeactivateLicenseRequestEntity$.outboundSchema = DeactivateLicenseRequestEntity$outboundSchema;
})(DeactivateLicenseRequestEntity$ || (DeactivateLicenseRequestEntity$ = {}));
export function deactivateLicenseRequestEntityToJSON(deactivateLicenseRequestEntity) {
    return JSON.stringify(DeactivateLicenseRequestEntity$outboundSchema.parse(deactivateLicenseRequestEntity));
}
export function deactivateLicenseRequestEntityFromJSON(jsonString) {
    return safeParse(jsonString, (x) => DeactivateLicenseRequestEntity$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'DeactivateLicenseRequestEntity' from JSON`);
}
