// creem/dist/esm/models/components/productlistentity.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
import * as z from "zod";
import { safeParse } from "../../lib/schemas.js";
import { PaginationEntity$inboundSchema, PaginationEntity$outboundSchema, } from "./paginationentity.js";
import { ProductEntity$inboundSchema, ProductEntity$outboundSchema, } from "./productentity.js";
/** @internal */
export const ProductListEntity$inboundSchema = z.object({
    items: z.array(ProductEntity$inboundSchema),
    pagination: PaginationEntity$inboundSchema,
});
/** @internal */
export const ProductListEntity$outboundSchema = z.object({
    items: z.array(ProductEntity$outboundSchema),
    pagination: PaginationEntity$outboundSchema,
});
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
export var ProductListEntity$;
(function (ProductListEntity$) {
    /** @deprecated use `ProductListEntity$inboundSchema` instead. */
    ProductListEntity$.inboundSchema = ProductListEntity$inboundSchema;
    /** @deprecated use `ProductListEntity$outboundSchema` instead. */
    ProductListEntity$.outboundSchema = ProductListEntity$outboundSchema;
})(ProductListEntity$ || (ProductListEntity$ = {}));
export function productListEntityToJSON(productListEntity) {
    return JSON.stringify(ProductListEntity$outboundSchema.parse(productListEntity));
}
export function productListEntityFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ProductListEntity$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ProductListEntity' from JSON`);
}
