// creem/dist/commonjs/models/operations/activatelicense.js
"use strict";
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActivateLicenseRequest$ = exports.ActivateLicenseRequest$outboundSchema = exports.ActivateLicenseRequest$inboundSchema = void 0;
exports.activateLicenseRequestToJSON = activateLicenseRequestToJSON;
exports.activateLicenseRequestFromJSON = activateLicenseRequestFromJSON;
const z = __importStar(require("zod"));
const primitives_js_1 = require("../../lib/primitives.js");
const schemas_js_1 = require("../../lib/schemas.js");
const components = __importStar(require("../components/index.js"));
/** @internal */
exports.ActivateLicenseRequest$inboundSchema = z.object({
    "x-api-key": z.string(),
    ActivateLicenseRequestEntity: components.ActivateLicenseRequestEntity$inboundSchema,
}).transform((v) => {
    return (0, primitives_js_1.remap)(v, {
        "x-api-key": "xApiKey",
        "ActivateLicenseRequestEntity": "activateLicenseRequestEntity",
    });
});
/** @internal */
exports.ActivateLicenseRequest$outboundSchema = z.object({
    xApiKey: z.string(),
    activateLicenseRequestEntity: components.ActivateLicenseRequestEntity$outboundSchema,
}).transform((v) => {
    return (0, primitives_js_1.remap)(v, {
        xApiKey: "x-api-key",
        activateLicenseRequestEntity: "ActivateLicenseRequestEntity",
    });
});
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
var ActivateLicenseRequest$;
(function (ActivateLicenseRequest$) {
    /** @deprecated use `ActivateLicenseRequest$inboundSchema` instead. */
    ActivateLicenseRequest$.inboundSchema = exports.ActivateLicenseRequest$inboundSchema;
    /** @deprecated use `ActivateLicenseRequest$outboundSchema` instead. */
    ActivateLicenseRequest$.outboundSchema = exports.ActivateLicenseRequest$outboundSchema;
})(ActivateLicenseRequest$ || (exports.ActivateLicenseRequest$ = ActivateLicenseRequest$ = {}));
function activateLicenseRequestToJSON(activateLicenseRequest) {
    return JSON.stringify(exports.ActivateLicenseRequest$outboundSchema.parse(activateLicenseRequest));
}
function activateLicenseRequestFromJSON(jsonString) {
    return (0, schemas_js_1.safeParse)(jsonString, (x) => exports.ActivateLicenseRequest$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ActivateLicenseRequest' from JSON`);
}
