// creem/dist/commonjs/models/operations/generatecustomerlinks.js
"use strict";
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.GenerateCustomerLinksRequest$ = exports.GenerateCustomerLinksRequest$outboundSchema = exports.GenerateCustomerLinksRequest$inboundSchema = void 0;
exports.generateCustomerLinksRequestToJSON = generateCustomerLinksRequestToJSON;
exports.generateCustomerLinksRequestFromJSON = generateCustomerLinksRequestFromJSON;
const z = __importStar(require("zod"));
const primitives_js_1 = require("../../lib/primitives.js");
const schemas_js_1 = require("../../lib/schemas.js");
const components = __importStar(require("../components/index.js"));
/** @internal */
exports.GenerateCustomerLinksRequest$inboundSchema = z.object({
    "x-api-key": z.string(),
    CreateCustomerPortalLinkRequestEntity: components.CreateCustomerPortalLinkRequestEntity$inboundSchema,
}).transform((v) => {
    return (0, primitives_js_1.remap)(v, {
        "x-api-key": "xApiKey",
        "CreateCustomerPortalLinkRequestEntity": "createCustomerPortalLinkRequestEntity",
    });
});
/** @internal */
exports.GenerateCustomerLinksRequest$outboundSchema = z.object({
    xApiKey: z.string(),
    createCustomerPortalLinkRequestEntity: components.CreateCustomerPortalLinkRequestEntity$outboundSchema,
}).transform((v) => {
    return (0, primitives_js_1.remap)(v, {
        xApiKey: "x-api-key",
        createCustomerPortalLinkRequestEntity: "CreateCustomerPortalLinkRequestEntity",
    });
});
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
var GenerateCustomerLinksRequest$;
(function (GenerateCustomerLinksRequest$) {
    /** @deprecated use `GenerateCustomerLinksRequest$inboundSchema` instead. */
    GenerateCustomerLinksRequest$.inboundSchema = exports.GenerateCustomerLinksRequest$inboundSchema;
    /** @deprecated use `GenerateCustomerLinksRequest$outboundSchema` instead. */
    GenerateCustomerLinksRequest$.outboundSchema = exports.GenerateCustomerLinksRequest$outboundSchema;
})(GenerateCustomerLinksRequest$ || (exports.GenerateCustomerLinksRequest$ = GenerateCustomerLinksRequest$ = {}));
function generateCustomerLinksRequestToJSON(generateCustomerLinksRequest) {
    return JSON.stringify(exports.GenerateCustomerLinksRequest$outboundSchema.parse(generateCustomerLinksRequest));
}
function generateCustomerLinksRequestFromJSON(jsonString) {
    return (0, schemas_js_1.safeParse)(jsonString, (x) => exports.GenerateCustomerLinksRequest$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'GenerateCustomerLinksRequest' from JSON`);
}
