// creem/dist/commonjs/models/operations/retrievecheckout.js
"use strict";
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.RetrieveCheckoutRequest$ = exports.RetrieveCheckoutRequest$outboundSchema = exports.RetrieveCheckoutRequest$inboundSchema = void 0;
exports.retrieveCheckoutRequestToJSON = retrieveCheckoutRequestToJSON;
exports.retrieveCheckoutRequestFromJSON = retrieveCheckoutRequestFromJSON;
const z = __importStar(require("zod"));
const primitives_js_1 = require("../../lib/primitives.js");
const schemas_js_1 = require("../../lib/schemas.js");
/** @internal */
exports.RetrieveCheckoutRequest$inboundSchema = z.object({
    checkout_id: z.string(),
    "x-api-key": z.string(),
}).transform((v) => {
    return (0, primitives_js_1.remap)(v, {
        "checkout_id": "checkoutId",
        "x-api-key": "xApiKey",
    });
});
/** @internal */
exports.RetrieveCheckoutRequest$outboundSchema = z.object({
    checkoutId: z.string(),
    xApiKey: z.string(),
}).transform((v) => {
    return (0, primitives_js_1.remap)(v, {
        checkoutId: "checkout_id",
        xApiKey: "x-api-key",
    });
});
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
var RetrieveCheckoutRequest$;
(function (RetrieveCheckoutRequest$) {
    /** @deprecated use `RetrieveCheckoutRequest$inboundSchema` instead. */
    RetrieveCheckoutRequest$.inboundSchema = exports.RetrieveCheckoutRequest$inboundSchema;
    /** @deprecated use `RetrieveCheckoutRequest$outboundSchema` instead. */
    RetrieveCheckoutRequest$.outboundSchema = exports.RetrieveCheckoutRequest$outboundSchema;
})(RetrieveCheckoutRequest$ || (exports.RetrieveCheckoutRequest$ = RetrieveCheckoutRequest$ = {}));
function retrieveCheckoutRequestToJSON(retrieveCheckoutRequest) {
    return JSON.stringify(exports.RetrieveCheckoutRequest$outboundSchema.parse(retrieveCheckoutRequest));
}
function retrieveCheckoutRequestFromJSON(jsonString) {
    return (0, schemas_js_1.safeParse)(jsonString, (x) => exports.RetrieveCheckoutRequest$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'RetrieveCheckoutRequest' from JSON`);
}
