// creem/dist/commonjs/models/operations/searchtransactions.js
"use strict";
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SearchTransactionsRequest$ = exports.SearchTransactionsRequest$outboundSchema = exports.SearchTransactionsRequest$inboundSchema = void 0;
exports.searchTransactionsRequestToJSON = searchTransactionsRequestToJSON;
exports.searchTransactionsRequestFromJSON = searchTransactionsRequestFromJSON;
const z = __importStar(require("zod"));
const primitives_js_1 = require("../../lib/primitives.js");
const schemas_js_1 = require("../../lib/schemas.js");
/** @internal */
exports.SearchTransactionsRequest$inboundSchema = z.object({
    customer_id: z.string().optional(),
    order_id: z.string().optional(),
    product_id: z.string().optional(),
    page_number: z.number().optional(),
    page_size: z.number().optional(),
    "x-api-key": z.string(),
}).transform((v) => {
    return (0, primitives_js_1.remap)(v, {
        "customer_id": "customerId",
        "order_id": "orderId",
        "product_id": "productId",
        "page_number": "pageNumber",
        "page_size": "pageSize",
        "x-api-key": "xApiKey",
    });
});
/** @internal */
exports.SearchTransactionsRequest$outboundSchema = z.object({
    customerId: z.string().optional(),
    orderId: z.string().optional(),
    productId: z.string().optional(),
    pageNumber: z.number().optional(),
    pageSize: z.number().optional(),
    xApiKey: z.string(),
}).transform((v) => {
    return (0, primitives_js_1.remap)(v, {
        customerId: "customer_id",
        orderId: "order_id",
        productId: "product_id",
        pageNumber: "page_number",
        pageSize: "page_size",
        xApiKey: "x-api-key",
    });
});
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
var SearchTransactionsRequest$;
(function (SearchTransactionsRequest$) {
    /** @deprecated use `SearchTransactionsRequest$inboundSchema` instead. */
    SearchTransactionsRequest$.inboundSchema = exports.SearchTransactionsRequest$inboundSchema;
    /** @deprecated use `SearchTransactionsRequest$outboundSchema` instead. */
    SearchTransactionsRequest$.outboundSchema = exports.SearchTransactionsRequest$outboundSchema;
})(SearchTransactionsRequest$ || (exports.SearchTransactionsRequest$ = SearchTransactionsRequest$ = {}));
function searchTransactionsRequestToJSON(searchTransactionsRequest) {
    return JSON.stringify(exports.SearchTransactionsRequest$outboundSchema.parse(searchTransactionsRequest));
}
function searchTransactionsRequestFromJSON(jsonString) {
    return (0, schemas_js_1.safeParse)(jsonString, (x) => exports.SearchTransactionsRequest$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'SearchTransactionsRequest' from JSON`);
}
