// creem/dist/commonjs/models/operations/searchproducts.js
"use strict";
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SearchProductsRequest$ = exports.SearchProductsRequest$outboundSchema = exports.SearchProductsRequest$inboundSchema = void 0;
exports.searchProductsRequestToJSON = searchProductsRequestToJSON;
exports.searchProductsRequestFromJSON = searchProductsRequestFromJSON;
const z = __importStar(require("zod"));
const primitives_js_1 = require("../../lib/primitives.js");
const schemas_js_1 = require("../../lib/schemas.js");
/** @internal */
exports.SearchProductsRequest$inboundSchema = z.object({
    page_number: z.number().optional(),
    page_size: z.number().optional(),
    "x-api-key": z.string(),
}).transform((v) => {
    return (0, primitives_js_1.remap)(v, {
        "page_number": "pageNumber",
        "page_size": "pageSize",
        "x-api-key": "xApiKey",
    });
});
/** @internal */
exports.SearchProductsRequest$outboundSchema = z.object({
    pageNumber: z.number().optional(),
    pageSize: z.number().optional(),
    xApiKey: z.string(),
}).transform((v) => {
    return (0, primitives_js_1.remap)(v, {
        pageNumber: "page_number",
        pageSize: "page_size",
        xApiKey: "x-api-key",
    });
});
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
var SearchProductsRequest$;
(function (SearchProductsRequest$) {
    /** @deprecated use `SearchProductsRequest$inboundSchema` instead. */
    SearchProductsRequest$.inboundSchema = exports.SearchProductsRequest$inboundSchema;
    /** @deprecated use `SearchProductsRequest$outboundSchema` instead. */
    SearchProductsRequest$.outboundSchema = exports.SearchProductsRequest$outboundSchema;
})(SearchProductsRequest$ || (exports.SearchProductsRequest$ = SearchProductsRequest$ = {}));
function searchProductsRequestToJSON(searchProductsRequest) {
    return JSON.stringify(exports.SearchProductsRequest$outboundSchema.parse(searchProductsRequest));
}
function searchProductsRequestFromJSON(jsonString) {
    return (0, schemas_js_1.safeParse)(jsonString, (x) => exports.SearchProductsRequest$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'SearchProductsRequest' from JSON`);
}
