// creem/src/models/operations/activatelicense.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */

import * as z from "zod";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { Result as SafeParseResult } from "../../types/fp.js";
import * as components from "../components/index.js";
import { SDKValidationError } from "../errors/sdkvalidationerror.js";

export type ActivateLicenseRequest = {
  xApiKey: string;
  activateLicenseRequestEntity: components.ActivateLicenseRequestEntity;
};

/** @internal */
export const ActivateLicenseRequest$inboundSchema: z.ZodType<
  ActivateLicenseRequest,
  z.ZodTypeDef,
  unknown
> = z.object({
  "x-api-key": z.string(),
  ActivateLicenseRequestEntity:
    components.ActivateLicenseRequestEntity$inboundSchema,
}).transform((v) => {
  return remap$(v, {
    "x-api-key": "xApiKey",
    "ActivateLicenseRequestEntity": "activateLicenseRequestEntity",
  });
});

/** @internal */
export type ActivateLicenseRequest$Outbound = {
  "x-api-key": string;
  ActivateLicenseRequestEntity:
    components.ActivateLicenseRequestEntity$Outbound;
};

/** @internal */
export const ActivateLicenseRequest$outboundSchema: z.ZodType<
  ActivateLicenseRequest$Outbound,
  z.ZodTypeDef,
  ActivateLicenseRequest
> = z.object({
  xApiKey: z.string(),
  activateLicenseRequestEntity:
    components.ActivateLicenseRequestEntity$outboundSchema,
}).transform((v) => {
  return remap$(v, {
    xApiKey: "x-api-key",
    activateLicenseRequestEntity: "ActivateLicenseRequestEntity",
  });
});

/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
export namespace ActivateLicenseRequest$ {
  /** @deprecated use `ActivateLicenseRequest$inboundSchema` instead. */
  export const inboundSchema = ActivateLicenseRequest$inboundSchema;
  /** @deprecated use `ActivateLicenseRequest$outboundSchema` instead. */
  export const outboundSchema = ActivateLicenseRequest$outboundSchema;
  /** @deprecated use `ActivateLicenseRequest$Outbound` instead. */
  export type Outbound = ActivateLicenseRequest$Outbound;
}

export function activateLicenseRequestToJSON(
  activateLicenseRequest: ActivateLicenseRequest,
): string {
  return JSON.stringify(
    ActivateLicenseRequest$outboundSchema.parse(activateLicenseRequest),
  );
}

export function activateLicenseRequestFromJSON(
  jsonString: string,
): SafeParseResult<ActivateLicenseRequest, SDKValidationError> {
  return safeParse(
    jsonString,
    (x) => ActivateLicenseRequest$inboundSchema.parse(JSON.parse(x)),
    `Failed to parse 'ActivateLicenseRequest' from JSON`,
  );
}
