// creem/src/models/operations/generatecustomerlinks.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */

import * as z from "zod";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { Result as SafeParseResult } from "../../types/fp.js";
import * as components from "../components/index.js";
import { SDKValidationError } from "../errors/sdkvalidationerror.js";

export type GenerateCustomerLinksRequest = {
  xApiKey: string;
  createCustomerPortalLinkRequestEntity:
    components.CreateCustomerPortalLinkRequestEntity;
};

/** @internal */
export const GenerateCustomerLinksRequest$inboundSchema: z.ZodType<
  GenerateCustomerLinksRequest,
  z.ZodTypeDef,
  unknown
> = z.object({
  "x-api-key": z.string(),
  CreateCustomerPortalLinkRequestEntity:
    components.CreateCustomerPortalLinkRequestEntity$inboundSchema,
}).transform((v) => {
  return remap$(v, {
    "x-api-key": "xApiKey",
    "CreateCustomerPortalLinkRequestEntity":
      "createCustomerPortalLinkRequestEntity",
  });
});

/** @internal */
export type GenerateCustomerLinksRequest$Outbound = {
  "x-api-key": string;
  CreateCustomerPortalLinkRequestEntity:
    components.CreateCustomerPortalLinkRequestEntity$Outbound;
};

/** @internal */
export const GenerateCustomerLinksRequest$outboundSchema: z.ZodType<
  GenerateCustomerLinksRequest$Outbound,
  z.ZodTypeDef,
  GenerateCustomerLinksRequest
> = z.object({
  xApiKey: z.string(),
  createCustomerPortalLinkRequestEntity:
    components.CreateCustomerPortalLinkRequestEntity$outboundSchema,
}).transform((v) => {
  return remap$(v, {
    xApiKey: "x-api-key",
    createCustomerPortalLinkRequestEntity:
      "CreateCustomerPortalLinkRequestEntity",
  });
});

/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
export namespace GenerateCustomerLinksRequest$ {
  /** @deprecated use `GenerateCustomerLinksRequest$inboundSchema` instead. */
  export const inboundSchema = GenerateCustomerLinksRequest$inboundSchema;
  /** @deprecated use `GenerateCustomerLinksRequest$outboundSchema` instead. */
  export const outboundSchema = GenerateCustomerLinksRequest$outboundSchema;
  /** @deprecated use `GenerateCustomerLinksRequest$Outbound` instead. */
  export type Outbound = GenerateCustomerLinksRequest$Outbound;
}

export function generateCustomerLinksRequestToJSON(
  generateCustomerLinksRequest: GenerateCustomerLinksRequest,
): string {
  return JSON.stringify(
    GenerateCustomerLinksRequest$outboundSchema.parse(
      generateCustomerLinksRequest,
    ),
  );
}

export function generateCustomerLinksRequestFromJSON(
  jsonString: string,
): SafeParseResult<GenerateCustomerLinksRequest, SDKValidationError> {
  return safeParse(
    jsonString,
    (x) => GenerateCustomerLinksRequest$inboundSchema.parse(JSON.parse(x)),
    `Failed to parse 'GenerateCustomerLinksRequest' from JSON`,
  );
}
