// creem/src/models/operations/retrievesubscription.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */

import * as z from "zod";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { Result as SafeParseResult } from "../../types/fp.js";
import { SDKValidationError } from "../errors/sdkvalidationerror.js";

export type RetrieveSubscriptionRequest = {
  /**
   * The unique identifier of the subscription
   */
  subscriptionId: string;
  xApiKey: string;
};

/** @internal */
export const RetrieveSubscriptionRequest$inboundSchema: z.ZodType<
  RetrieveSubscriptionRequest,
  z.ZodTypeDef,
  unknown
> = z.object({
  subscription_id: z.string(),
  "x-api-key": z.string(),
}).transform((v) => {
  return remap$(v, {
    "subscription_id": "subscriptionId",
    "x-api-key": "xApiKey",
  });
});

/** @internal */
export type RetrieveSubscriptionRequest$Outbound = {
  subscription_id: string;
  "x-api-key": string;
};

/** @internal */
export const RetrieveSubscriptionRequest$outboundSchema: z.ZodType<
  RetrieveSubscriptionRequest$Outbound,
  z.ZodTypeDef,
  RetrieveSubscriptionRequest
> = z.object({
  subscriptionId: z.string(),
  xApiKey: z.string(),
}).transform((v) => {
  return remap$(v, {
    subscriptionId: "subscription_id",
    xApiKey: "x-api-key",
  });
});

/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
export namespace RetrieveSubscriptionRequest$ {
  /** @deprecated use `RetrieveSubscriptionRequest$inboundSchema` instead. */
  export const inboundSchema = RetrieveSubscriptionRequest$inboundSchema;
  /** @deprecated use `RetrieveSubscriptionRequest$outboundSchema` instead. */
  export const outboundSchema = RetrieveSubscriptionRequest$outboundSchema;
  /** @deprecated use `RetrieveSubscriptionRequest$Outbound` instead. */
  export type Outbound = RetrieveSubscriptionRequest$Outbound;
}

export function retrieveSubscriptionRequestToJSON(
  retrieveSubscriptionRequest: RetrieveSubscriptionRequest,
): string {
  return JSON.stringify(
    RetrieveSubscriptionRequest$outboundSchema.parse(
      retrieveSubscriptionRequest,
    ),
  );
}

export function retrieveSubscriptionRequestFromJSON(
  jsonString: string,
): SafeParseResult<RetrieveSubscriptionRequest, SDKValidationError> {
  return safeParse(
    jsonString,
    (x) => RetrieveSubscriptionRequest$inboundSchema.parse(JSON.parse(x)),
    `Failed to parse 'RetrieveSubscriptionRequest' from JSON`,
  );
}
