// creem/src/models/operations/searchtransactions.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */

import * as z from "zod";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { Result as SafeParseResult } from "../../types/fp.js";
import { SDKValidationError } from "../errors/sdkvalidationerror.js";

export type SearchTransactionsRequest = {
  /**
   * The customer id
   */
  customerId?: string | undefined;
  /**
   * The order id
   */
  orderId?: string | undefined;
  /**
   * The product id
   */
  productId?: string | undefined;
  /**
   * The page number
   */
  pageNumber?: number | undefined;
  /**
   * The the page size
   */
  pageSize?: number | undefined;
  xApiKey: string;
};

/** @internal */
export const SearchTransactionsRequest$inboundSchema: z.ZodType<
  SearchTransactionsRequest,
  z.ZodTypeDef,
  unknown
> = z.object({
  customer_id: z.string().optional(),
  order_id: z.string().optional(),
  product_id: z.string().optional(),
  page_number: z.number().optional(),
  page_size: z.number().optional(),
  "x-api-key": z.string(),
}).transform((v) => {
  return remap$(v, {
    "customer_id": "customerId",
    "order_id": "orderId",
    "product_id": "productId",
    "page_number": "pageNumber",
    "page_size": "pageSize",
    "x-api-key": "xApiKey",
  });
});

/** @internal */
export type SearchTransactionsRequest$Outbound = {
  customer_id?: string | undefined;
  order_id?: string | undefined;
  product_id?: string | undefined;
  page_number?: number | undefined;
  page_size?: number | undefined;
  "x-api-key": string;
};

/** @internal */
export const SearchTransactionsRequest$outboundSchema: z.ZodType<
  SearchTransactionsRequest$Outbound,
  z.ZodTypeDef,
  SearchTransactionsRequest
> = z.object({
  customerId: z.string().optional(),
  orderId: z.string().optional(),
  productId: z.string().optional(),
  pageNumber: z.number().optional(),
  pageSize: z.number().optional(),
  xApiKey: z.string(),
}).transform((v) => {
  return remap$(v, {
    customerId: "customer_id",
    orderId: "order_id",
    productId: "product_id",
    pageNumber: "page_number",
    pageSize: "page_size",
    xApiKey: "x-api-key",
  });
});

/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
export namespace SearchTransactionsRequest$ {
  /** @deprecated use `SearchTransactionsRequest$inboundSchema` instead. */
  export const inboundSchema = SearchTransactionsRequest$inboundSchema;
  /** @deprecated use `SearchTransactionsRequest$outboundSchema` instead. */
  export const outboundSchema = SearchTransactionsRequest$outboundSchema;
  /** @deprecated use `SearchTransactionsRequest$Outbound` instead. */
  export type Outbound = SearchTransactionsRequest$Outbound;
}

export function searchTransactionsRequestToJSON(
  searchTransactionsRequest: SearchTransactionsRequest,
): string {
  return JSON.stringify(
    SearchTransactionsRequest$outboundSchema.parse(searchTransactionsRequest),
  );
}

export function searchTransactionsRequestFromJSON(
  jsonString: string,
): SafeParseResult<SearchTransactionsRequest, SDKValidationError> {
  return safeParse(
    jsonString,
    (x) => SearchTransactionsRequest$inboundSchema.parse(JSON.parse(x)),
    `Failed to parse 'SearchTransactionsRequest' from JSON`,
  );
}
